diff --git a/utilities/ovs-dev.py b/utilities/ovs-dev.py
index 9ce0f04c7..f245a0404 100755
--- a/utilities/ovs-dev.py
+++ b/utilities/ovs-dev.py
@@ -284,9 +284,16 @@ def run():
             "other_config:dpdk-init=true" % root_uuid)
         _sh("ovs-vsctl --no-wait set Open_vSwitch %s other_config:"
             "dpdk-extra=\"%s\"" % (root_uuid, ' '.join(options.dpdk)))
+        dpdk_version = _sh("ovs-vswitchd --version", capture=True)
+        dpdk_version = [s for s in version if "DPDK" in s]
+        if dpdk_version:
+            dpdk_version = dpdk_version[0].decode().strip().split()[1]
+            _sh("ovs-vsctl --no-wait set Open_vSwitch %sdpdk_version=%s"
+                % (root_uuid, dpdk_version))
     else:
         _sh("ovs-vsctl --no-wait set Open_vSwitch %s "
             "other_config:dpdk-init=false" % root_uuid)
+        _sh("ovs-vsctl --no-wait clear Open_vSwitch %s dpdk_version" % root_uuid)
 
     if options.gdb:
         cmd = ["gdb", "--args"] + cmd