Spell/Hunter: Camouflage (Pet)

diff -r 786eaa087c70 -r aaae93771646 src/server/game/Spells/SpellEffects.cpp
--- a/src/server/game/Spells/SpellEffects.cpp	Sat Sep 17 00:09:07 2011 +0400
+++ b/src/server/game/Spells/SpellEffects.cpp	Sat Sep 17 09:57:12 2011 +0400
@@ -1447,6 +1447,9 @@
                 m_caster->CastSpell(m_caster, 77443, true);
             if (m_spellInfo->SpellFamilyFlags[2] & 0x20)
                 m_caster->CastSpell(m_caster, 51755, true);
+            if (m_spellInfo->Id == 51753)
+                if (Unit* pet = m_caster->GetGuardianPet())
+                    pet->CastSpell(pet, 51753, true);
             break;
         case SPELLFAMILY_PRIEST:
         {