diff --git a/gcc/dumpfile.c b/gcc/dumpfile.c
index 658500b6103..a5fbc77a30f 100644
--- a/gcc/dumpfile.c
+++ b/gcc/dumpfile.c
@@ -798,6 +798,7 @@ dump_switch_p_1 (const char *arg, struct dump_file_info *dfi, bool doglob)
   ptr = option_value;
   flags = 0;
 
+  int pstate = -1;
   while (*ptr)
     {
       const struct dump_option_value_info *option_ptr;
@@ -832,6 +833,7 @@ dump_switch_p_1 (const char *arg, struct dump_file_info *dfi, bool doglob)
           if (dfi->pfilename)
             free (CONST_CAST (char *, dfi->pfilename));
           dfi->pfilename = xstrdup (ptr + 1);
+         pstate = 1;
           break;
         }
       else
@@ -841,7 +843,7 @@ dump_switch_p_1 (const char *arg, struct dump_file_info *dfi, bool doglob)
       ptr = end_ptr;
     }
 
-  dfi->pstate = -1;
+  dfi->pstate = pstate;
   dfi->pflags |= flags;
 
   /* Process -fdump-tree-all and -fdump-rtl-all, by enabling all the