diff -r 046320bc0f82 MoinMoin/config/default.py
--- a/MoinMoin/config/default.py        Sat Oct 08 21:12:52 2011 +0200
+++ b/MoinMoin/config/default.py        Sat Oct 08 22:09:37 2011 +0200
@@ -555,7 +555,8 @@
         'These settings control outgoing and incoming email from and to the wiki.',
     (
       ('from', None, "Used as From: address for generated mail. [Unicode]"),
-      ('login', None, "'username userpass' for SMTP server authentication (None = don't use auth)."),
+      ('username', None, "Username for SMTP server authentication (None = don't use auth)."),
+      ('password', None, "Password for SMTP server authentication (None = don't use auth)."),
       ('smarthost', None, "Address of SMTP server to use for sending mail (None = don't use SMTP server)."),
       ('sendmail', None, "sendmail command to use for sending mail (None = don't use sendmail)"),
     )),
diff -r 046320bc0f82 MoinMoin/mail/sendmail.py
--- a/MoinMoin/mail/sendmail.py Sat Oct 08 21:12:52 2011 +0200
+++ b/MoinMoin/mail/sendmail.py Sat Oct 08 22:09:37 2011 +0200
@@ -133,8 +133,7 @@
             server = smtplib.SMTP(host, int(port))
             try:
                 #server.set_debuglevel(1)
-                if cfg.mail_login:
-                    user, pwd = cfg.mail_login.split()
+                if cfg.mail_username is not None and cfg.mail_password is not None:
                     try: # try to do tls
                         server.ehlo()
                         if server.has_extn('starttls'):
@@ -144,7 +143,7 @@
                     except:
                         logging.debug("could not establish a tls connection to smtp server, continuing without tls")
                     logging.debug("trying to log in to smtp server using account '%s'" % user)
-                    server.login(user, pwd)
+                    server.login(cfg.mail_username, cfg.mail_password)
                 server.sendmail(mail_from, to, msg.as_string())
             finally:
                 try: