--- time.rpy.old	2015-08-16 00:51:55.000000000 -0300
+++ time.rpy	2015-08-16 00:53:38.000000000 -0300
@@ -342,11 +342,11 @@
 
     # a "Il aurait pourtant été beaucoup plus sûr de demander à ce qu'on m'envoie un matériel de rechange plutôt que de me soumettre aux caprices d'un régulateur totalement obsolète et sans doute aléatoire. Tant pis !"
     a "It would have been a lot safer to ask that spare equipment would be sent to me rather than rely on an outdated and probably uncertain regulator. So be it!"
     
 # game/time.rpy:305
-translate en histoire_time_7f8c2024:
+translate en histoire_time_718ef2e7:
 
     # nvl clear
     # a "Par contre, sans interface de communication avec la Tour, la synchronisation temporelle risque d'être vraiment délicate...{w} Il va falloir que je calcule manuellement le timecode de retour, en espérant que l'appareil me permettre de l'entrer avec suffisamment de précision."
     nvl clear
     a "What's bad is that without any communication-interface with the Tower, time synchronization is going to be awfully tricky... I will have to compute the return timecode by hand. Let's hope I will be able to enter it into the device with enough precision."
@@ -362,11 +362,11 @@
 
     # a "J'aurais presque envie de faire un essai préalable, sur une époque plus proche, mais il s'agirait d'une translation illégale."
     a "I almost feel like testing it first, into a closer time, but it would be totally inappropriate: that would be an illegal translation."
 
 # game/time.rpy:311
-translate en histoire_time_85bcc7f9:
+translate en histoire_time_e4b45c8a:
 
     # nvl clear
     # a "Même si, avec mon expérience d'Agent Voyageur, il n'y aurait que très peu de chance que je cause une interférence ou que je me fasse surprendre, je ne peux pas me résoudre à violer la Charte, même pour sauver ma propre vie."
     nvl clear
     a "I know that, considering my Agent Traveller skills, I probably would not cause any interference, even less be caught; though I cannot resolve to violate the Code, even to save my own life."