Common subdirectories: a/mg-20171014/CVS and b/mg-20171014/CVS
diff -upN a/mg-20171014/def.h b/mg-20171014/def.h
--- a/mg-20171014/def.h	2018-01-27 22:27:04.798214738 +0900
+++ b/mg-20171014/def.h	2018-01-27 22:30:13.154577600 +0900
@@ -454,6 +454,7 @@ void		vttidy(void);
 void		update(int);
 int		linenotoggle(int, int);
 int		colnotoggle(int, int);
+int		onelnscroll(int, int);
 
 /* echo.c X */
 void		 eerase(void);
diff -upN a/mg-20171014/display.c b/mg-20171014/display.c
--- a/mg-20171014/display.c	2018-01-27 22:27:04.798214738 +0900
+++ b/mg-20171014/display.c	2018-01-27 22:30:13.154577600 +0900
@@ -92,6 +92,7 @@ struct score *score;			/* [NROW * NROW]
 
 static int	 linenos = TRUE;
 static int	 colnos = FALSE;
+static int	 onelinescroll = FALSE;
 
 /* Is macro recording enabled? */
 extern int macrodef;
@@ -130,6 +131,20 @@ colnotoggle(int f, int n)
 	return (TRUE);
 }
 
+/* ARGSUSED */
+int
+onelnscroll(int f, int n)
+{
+	if (f & FFARG)
+		onelinescroll = n > 0;
+	else
+		onelinescroll = !onelinescroll;
+		
+	ewprintf("one-line-scroll was %sabled", onelinescroll ? "en" : "dis");
+
+	return (TRUE);
+}
+
 /*
  * Reinit the display data structures, this is called when the terminal
  * size changes.
@@ -461,6 +476,12 @@ update(int modelinecolor)
 			i += wp->w_ntrows;
 			if (i < 0)
 				i = 0;
+		} else if (onelinescroll && (thisflag & CFCPCN)) {
+			i = 0;
+			lp = wp->w_linep;
+			if (lback(lp) != wp->w_dotp)
+				while ((lp = lforw(lp)) != wp->w_dotp)
+					++i;
 		} else
 			i = wp->w_ntrows / 2; /* current center, no change */
 
diff -upN a/mg-20171014/funmap.c b/mg-20171014/funmap.c
--- a/mg-20171014/funmap.c	2018-01-27 22:27:04.798214738 +0900
+++ b/mg-20171014/funmap.c	2018-01-27 22:30:13.154577600 +0900
@@ -216,6 +216,7 @@ static struct funmap functnames[] = {
 	{showcpos, "what-cursor-position",},
 	{filewrite, "write-file",},
 	{yank, "yank",},
+	{onelnscroll, "one-line-scroll"},
 	{NULL, NULL,}
 };
 
diff -upN a/mg-20171014/match.c b/mg-20171014/match.c
--- a/mg-20171014/match.c	2018-01-27 22:27:04.798214738 +0900
+++ b/mg-20171014/match.c	2018-01-27 22:31:34.472601770 +0900
@@ -156,7 +156,7 @@ displaymatch(struct line *clp, int cbo)
 		curwp->w_rflag |= WFMOVE;
 
 		update(CMODE);		/* show match */
-		ttwait(1000);		/* wait for key or 1 second */
+		ttwait(300);		/* wait for key or 0.3 second */
 
 		curwp->w_dotp = tlp;	/* return to old position */
 		curwp->w_doto = tbo;