ctive_handler();
diff --git a/drupal/profiles/mal/modules/contrib/search_api/contrib/search_api_facetapi/plugins/facetapi/query_type_term.inc b/drupal/profiles/mal/modules/contrib/search_api/contrib/search_api_facetapi/plugins/facetapi/query_type_term.inc
index b2c02c3..58d47a4 100644
--- a/drupal/profiles/mal/modules/contrib/search_api/contrib/search_api_facetapi/plugins/facetapi/query_type_term.inc
+++ b/drupal/profiles/mal/modules/contrib/search_api/contrib/search_api_facetapi/plugins/facetapi/query_type_term.inc
@@ -109,6 +109,11 @@ class SearchApiFacetapiTerm extends FacetapiQueryType implements FacetapiQueryTy
         if ($upper != '*') {
           $query_filter->condition($field, $upper, '<=');
         }
+        // todo add settings
+        if (true) {
+          $query_filter->condition($field, $upper, '<>>');
+          $query_filter->condition($field, $lower, '<><');
+        }
       }
       else {
         // Same as above, but with inverted logic.
diff --git a/drupal/profiles/mal/modules/contrib/search_api_solr/includes/service.inc b/drupal/profiles/mal/modules/contrib/search_api_solr/includes/service.inc
index 76225fa..8c4f6cc 100644
--- a/drupal/profiles/mal/modules/contrib/search_api_solr/includes/service.inc
+++ b/drupal/profiles/mal/modules/contrib/search_api_solr/includes/service.inc
@@ -1596,6 +1596,10 @@ class SearchApiSolrService extends SearchApiAbstractService {
         return "$field:[$value TO *]";
       case '>':
         return "$field:{{$value} TO *}";
+      case '<>>':
+        return "*:* AND -($field:{{$value} TO *})";
+      case '<><':
+        return "*:* AND -($field:{* TO $value})";
 
       default:
         return "$field:$value";