diff --git a/tree_models.lua b/tree_models.lua
index f1cf5a5..bbe9856 100644
--- a/tree_models.lua
+++ b/tree_models.lua
@@ -144,8 +144,8 @@ moretrees.pine_model={
 	rules_b="[&&&TTT[++^Fdd][--&Fdd]//dd[+^d][--&Fd]]",
 	rules_c="/",
 	rules_d="F",
-	trunk="moretrees:pine_trunk",
-	leaves="moretrees:pine_leaves",
+	trunk="default:pine_tree",
+	leaves="default:pine_needles",
 	angle=30,
 	iterations=2,
 	random_level=0,
@@ -191,8 +191,8 @@ moretrees.acacia_model={
 			.."GGfffff--G--"
 			.."ffffGGG",
 	rules_c = "/",
-	trunk="moretrees:acacia_trunk",
-	leaves="moretrees:acacia_leaves",
+	trunk="default:acacia_tree",
+	leaves="default:acacia_wood",
 	angle=45,
 	iterations=3,
 	random_level=0,