diff --git a/libhb/nlmeans.c b/libhb/nlmeans.c
index 3e78ab9..a306c8f 100644
--- a/libhb/nlmeans.c
+++ b/libhb/nlmeans.c
@@ -55,8 +55,8 @@
 #include "taskset.h"
 #include "nlmeans.h"
 
-#define NLMEANS_STRENGTH_LUMA_DEFAULT      8
-#define NLMEANS_STRENGTH_CHROMA_DEFAULT    8
+#define NLMEANS_STRENGTH_LUMA_DEFAULT      6
+#define NLMEANS_STRENGTH_CHROMA_DEFAULT    6
 #define NLMEANS_ORIGIN_TUNE_LUMA_DEFAULT   1
 #define NLMEANS_ORIGIN_TUNE_CHROMA_DEFAULT 1
 #define NLMEANS_PATCH_SIZE_LUMA_DEFAULT    7
diff --git a/test/test.c b/test/test.c
index 513add9..7aa8aad 100644
--- a/test/test.c
+++ b/test/test.c
@@ -2204,7 +2204,7 @@ static int ParseOptions( int argc, char ** argv )
                 }
                 else
                 {
-                    nlmeans = strdup("light");
+                    nlmeans = strdup("medium");
                 }
                 break;
             case FILTER_NLMEANS_TUNE: