diff --git a/epg_ui/source/prod_epg/core_services/src/services_model/NetworkModel.cpp b/epg_ui/source/prod_epg/core_services/src/services_model/NetworkModel.cpp
index ba42dab..b9af9c1 100644
--- a/epg_ui/source/prod_epg/core_services/src/services_model/NetworkModel.cpp
+++ b/epg_ui/source/prod_epg/core_services/src/services_model/NetworkModel.cpp
@@ -351,6 +351,9 @@ void NetworkModel::networkStatusHandler(JsonResponse *response, JsonEventsReques
 
     Q_UNUSED(pRequest)
 
+    qDebug() << "networkStatusHandler:" << response->getBody();
+    qDebug() << "networkStatusHandler:" << response->m_result ;
+
     processNetworkStatusUpdate(response->m_result);
 
 }
@@ -1001,7 +1004,9 @@ void NetworkModel::startWirelessNetworkScan()
 
         // std::cout << "startWirelessNetworkScanTimer: ----- " << std::endl;
 
-        JsonRequest *pRequest = RequestFactory::instance().createRestRequest(amsUri, resUri, "" /*queryUri*/, JsonRequest::HTTP_METHOD_POST); // POST
+        //JsonRequest *pRequest = RequestFactory::instance().createRestRequest(amsUri, resUri, "" /*queryUri*/, JsonRequest::HTTP_METHOD_POST); // POST
+        qDebug() << "Sending scanstart. URL:" << "ams://network1" << resUri;
+        JsonRequest *pRequest = RequestFactory::instance().createRestRequest("ams://network1", resUri, "" /*queryUri*/, JsonRequest::HTTP_METHOD_POST); // POST
         SAFE_CONNECT_UNIQUE(pRequest, &JsonRequest::responseReceived, this, &NetworkModel::wirelessNetworkScanRequestCallback);
         pRequest->send();
     }
@@ -1093,7 +1098,9 @@ void NetworkModel::networkWirelessScanHandler(JsonResponse *response, JsonEvents
 
     // std::cout << "networkWirelessScanHandler: ----- " << response->getBody().toStdString().c_str() << std::endl;
 
-    Q_UNUSED(pRequest)
+    Q_UNUSED(pRequest);
+
+    qWarning() <<"XXXYYY:networkWirelessScanHandler"<< response->getBody();
 
     processNetworkWirelessScanUpdate(response->m_result);