diff --git a/gst/gstbuffer.c b/gst/gstbuffer.c
index 34ca517..fc19de3 100644
--- a/gst/gstbuffer.c
+++ b/gst/gstbuffer.c
@@ -234,9 +234,13 @@ _get_merged_memory (GstBuffer * buffer, guint idx, guint length)
     } else {
       gsize i, tocopy, left;
       GstMapInfo sinfo, dinfo;
+      GstAllocationParams params = { 0, 0, 0, 0, };
       guint8 *ptr;
 
-      result = gst_allocator_alloc (NULL, size, NULL);
+      for (i = 0; i < length; i++)
+        params.align |= mem[i]->align;
+
+      result = gst_allocator_alloc (NULL, size, &params);