diff --git a/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResResSpec.java b/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResResSpec.java
index 645821fd..37bf3569 100755
--- a/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResResSpec.java
+++ b/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResResSpec.java
@@ -113,7 +113,7 @@ public class ResResSpec {
     }
 
     public void addResource(ResResource res) throws AndrolibException {
-        addResource(res, false);
+        addResource(res, true);
     }
 
     public void addResource(ResResource res, boolean overwrite) throws AndrolibException {
diff --git a/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResType.java b/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResType.java
index cfe45656..94db811d 100644
--- a/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResType.java
+++ b/brut.apktool/apktool-lib/src/main/java/brut/androlib/res/data/ResType.java
@@ -52,7 +52,7 @@ public class ResType {
     }
 
     public void addResource(ResResource res) throws AndrolibException {
-        addResource(res, false);
+        addResource(res, true);
     }
 
     public void removeResource(ResResource res) throws AndrolibException {