diff --git a/page_manager/theme/page-manager-search-results.tpl.php b/page_manager/theme/page-manager-search-results.tpl.php
index 1933cb9..2bfa6eb 100644
--- a/page_manager/theme/page-manager-search-results.tpl.php
+++ b/page_manager/theme/page-manager-search-results.tpl.php
@@ -11,7 +11,5 @@
  */
 ?>
 <<?php print $list_type; ?> class="search-results <?php print $module; ?>-results">
-  <?php foreach ($results as $result) : ?>
-    <?php print theme('search_result', array('result' => $result, 'module' => $module)); ?>
-  <?php endforeach; ?>
+  <?php print $results; ?>
 </<?php print $list_type; ?>>
diff --git a/plugins/content_types/search/search_result.inc b/plugins/content_types/search/search_result.inc
index 9559ac2..189d065 100644
--- a/plugins/content_types/search/search_result.inc
+++ b/plugins/content_types/search/search_result.inc
@@ -85,7 +85,11 @@ function ctools_search_result_content_type_render($subtype, $conf, $panel_args,
   }

   if (!empty($results['#results'])) {
-    $output = theme('page_manager_search_results', array('results' => $results['#results'], 'module' => $conf['type']));
+    $themed_results = array();
+    foreach ($results['#results'] as $result) {
+      $themed_results[] = theme('search_result', array('result' => $result, 'module' => $conf['type']));
+    }
+    $output = theme('page_manager_search_results', array('results' => implode('', $themed_results), 'module' => $conf['type']));
     $output .= theme('pager', array('tags' => NULL));