diff --git a/lib/ftp.c b/lib/ftp.c
index fade092..ad18d45 100644
--- a/lib/ftp.c
+++ b/lib/ftp.c
@@ -338,10 +338,11 @@ static CURLcode AcceptServerConnect(struct connectdata *conn)
   if(CURL_SOCKET_BAD == s) {
     failf(data, "Error accept()ing server connect");
     return CURLE_FTP_PORT_FAILED;
   }
   infof(data, "Connection accepted from server\n");
+  conn->bits.do_more = FALSE;
 
   conn->sock[SECONDARYSOCKET] = s;
   (void)curlx_nonblock(s, TRUE); /* enable non-blocking */
   conn->sock_accepted[SECONDARYSOCKET] = TRUE;