 src/map/clif.cpp   |  8 ++++----
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/src/map/clif.cpp b/src/map/clif.cpp
index 6efe626..fc6f110 100644
--- a/src/map/clif.cpp
+++ b/src/map/clif.cpp
@@ -13295,10 +13295,10 @@ void clif_parse_CreateGuild(int fd,struct map_session_data *sd){
 		return;
 	}
 
-	if(sd->clan){
+	/*if(sd->clan){
 		// Should display a clientside message "You are currently joined in Clan !!" so we ignore it
 		return;
-	}
+	}*/
 
 	guild_create(sd, name);
 }
@@ -13527,9 +13527,9 @@ int clif_sub_guild_invite(int fd, struct map_session_data *sd, struct map_sessio
 	}
 
 	// Players in a clan can not join a guild
-	if(t_sd && t_sd->clan){
+	/*if(t_sd && t_sd->clan){
 		return 1;
-	}
+	}*/
 
 	guild_invite(sd, t_sd);
 	return 0;