diff --git a/debian/enc-poller-amedia.init b/debian/enc-poller-amedia.init
index a702659..e4749dd 100644
--- a/debian/enc-poller-amedia.init
+++ b/debian/enc-poller-amedia.init
@@ -55,7 +55,7 @@ do_start()
                || return 1
         start-stop-daemon --start --quiet --make-pidfile --no-close --background \
                  --pidfile $PIDFILE --chuid ${USER} --exec $DAEMON -- \
-                $DAEMON_ARGS 2>>/var/log/enc/poller/${NAME}.panic.log 1>>/dev/null \
+                $DAEMON_ARGS 2>>/var/log/enc-poller-amedia/${NAME}.panic.log 1>>/dev/null \
                 || return 2
        # Add code here, if necessary, that waits for the process to be ready
        # to handle requests from services started subsequently which depend
diff --git a/debian/enc-worker-amedia.init b/debian/enc-worker-amedia.init
index 0d7bf53..e4c87a6 100644
--- a/debian/enc-worker-amedia.init
+++ b/debian/enc-worker-amedia.init
@@ -55,7 +55,7 @@ do_start()
                || return 1
         start-stop-daemon --start --quiet --make-pidfile --no-close --background \
                  --pidfile $PIDFILE --chuid ${USER} --exec $DAEMON -- \
-                $DAEMON_ARGS 2>>/var/log/enc/poller/${NAME}.panic.log 1>>/dev/null \
+                $DAEMON_ARGS 2>>/var/log/enc-worker-amedia/${NAME}.panic.log 1>>/dev/null \
                 || return 2
        # Add code here, if necessary, that waits for the process to be ready
        # to handle requests from services started subsequently which depend