diff --git a/src/gui/movieplayer.cpp b/src/gui/movieplayer.cpp
index ebc0835..4b45739 100644
--- a/src/gui/movieplayer.cpp
+++ b/src/gui/movieplayer.cpp
@@ -1757,9 +1757,11 @@ void CMoviePlayerGui::PlayFileLoop(void)
                        //NI if (timeshift == TSHIFT_MODE_OFF)
                                callInfoViewer();
                } else if (msg == (neutrino_msg_t) g_settings.mpkey_bookmark) {
+#if HAVE_COOL_HARDWARE
                        if (is_file_player)
                                selectChapter();
                        else
+#endif
                                handleMovieBrowser((neutrino_msg_t) g_settings.mpkey_bookmark, position);
                        update_lcd = true;
                        clearSubtitle();