diff --git a/src/engine/qcommon/md5.c b/src/engine/qcommon/md5.c
index 9c476ee..df4521e 100644
--- a/src/engine/qcommon/md5.c
+++ b/src/engine/qcommon/md5.c
@@ -370,7 +370,7 @@ static char *CalculateGUID(const char *key)
 
 char           *Com_MD5FileETCompat(const char *filename)
 {
-       char                       key[18];
+       char                       key[19] = {0};
        char              *buffer;
        int                len;
 
@@ -382,7 +382,6 @@ char           *Com_MD5FileETCompat(const char *filename)
                {
                        key[i] = buffer[i + 10];
                }
-               key[18] = 0;
 
                return CalculateGUID(key);
        }