vanhofen@jessie:~/ni/buildsystem[master]$ git diff
diff --git a/make/ffmpeg-arm.mk b/make/ffmpeg-arm.mk
index 908b13e..25cc589 100644
--- a/make/ffmpeg-arm.mk
+++ b/make/ffmpeg-arm.mk
@@ -25,7 +25,7 @@ FFMPEG_PATCH += ffmpeg-$(FFMPEG_VER)-chunked_transfer_fix_eof.patch

 # -----------------------------------------------------------------------------

-FFMPEG_DEPS = $(D)/openssl $(D)/freetype $(D)/librtmp $(D)/libbluray $(D)/libass $(D)/libxml2
+FFMPEG_DEPS = $(D)/openssl $(D)/freetype $(D)/librtmp $(D)/libbluray $(D)/libass $(D)/libxml2 $(D)/alsa-lib

 # -----------------------------------------------------------------------------

diff --git a/make/neutrino.mk b/make/neutrino.mk
index cbe15de..c890a6b 100644
--- a/make/neutrino.mk
+++ b/make/neutrino.mk
@@ -2,7 +2,7 @@

 N_DEPS = libcurl freetype libjpeg giflib ffmpeg openthreads openssl libdvbsi ntp libsigc++ luaposix pugixml libfribidi

-LH_DEPS = ffmpeg alsa-lib
+LH_DEPS = ffmpeg

 ifeq ($(BOXTYPE)-$(HAS_LIBCS), coolstream-yes)
        N_DEPS += libcoolstream