diff --git a/5.5/Dockerfile b/5.5/Dockerfile
index af4df75..484e43f 100644
--- a/5.5/Dockerfile
+++ b/5.5/Dockerfile
@@ -29,7 +29,8 @@ COPY contrib /var/lib/mysql/
 # safe in the future. This should *never* change, the last test is there
 # to make sure of that.
 RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
-    yum -y --setopt=tsflags=nodocs install https://www.softwarecollections.org/en/scls/rhscl/mysql55/epel-7-x86_64/download/rhscl-mysql55-epel-7-x86_64.noarch.rpm && \
+    yum -y --setopt=tsflags=nodocs install wget && \
+    wget https://copr.fedoraproject.org/coprs/rhscl/mysql55-el7/repo/epel-7/rhscl-mysql55-el7-epel-7.repo -O /etc/yum.repos.d/rhscl-mysql55-el7-epel-7.repo && \
     yum -y --setopt=tsflags=nodocs install gettext hostname bind-utils mysql55 && \
     yum clean all && \
     mkdir -p /var/lib/mysql/data && chown mysql.mysql /var/lib/mysql/data && \