diff --git a/lib/upipe/uprobe_stdio.c b/lib/upipe/uprobe_stdio.c
index 76ba8ce6..98e85590 100644
--- a/lib/upipe/uprobe_stdio.c
+++ b/lib/upipe/uprobe_stdio.c
@@ -38,6 +38,7 @@
 #include <stdio.h>
 #include <string.h>
 #include <stdarg.h>
+#include <sys/time.h>
 
 /** maximum level name length */
 #define LEVEL_NAME_LEN  7
@@ -179,14 +180,18 @@ static int uprobe_stdio_throw(struct uprobe *uprobe, struct upipe *upipe,
 
     const struct level *level = uprobe_stdio_get_level(ulog);
 
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    uint64_t now = tv.tv_sec * 1000000 + tv.tv_usec;
+
     if (colored)
         fprintf(uprobe_stdio->stream,
-                LEVEL_COLOR("%s", "%*s") ": %s%s\n",
-                level->color, LEVEL_NAME_LEN, level->name, buffer, ulog->msg);
+                LEVEL_COLOR("%s", "%*s") ": %" PRIu64 " %s%s\n",
+                level->color, LEVEL_NAME_LEN, level->name, now, buffer, ulog->msg);
     else
         fprintf(uprobe_stdio->stream,
-                LEVEL_NOCOLOR("%s") ": %s%s\n",
-                level->name, buffer, ulog->msg);
+                LEVEL_NOCOLOR("%s") ": %" PRIu64 " %s%s\n",
+                level->name, now, buffer, ulog->msg);
 
     return UBASE_ERR_NONE;
 }
stash@{8}: WIP on dvbsrc: 6d7231c9 dvbsrc: update to new version of patch