Spell/Paladin: Templar's Verdict

diff -r 309c073286cd -r 56e5f4229fd4 src/server/game/Spells/SpellEffects.cpp
--- a/src/server/game/Spells/SpellEffects.cpp	Sat Sep 17 09:59:42 2011 +0400
+++ b/src/server/game/Spells/SpellEffects.cpp	Sat Sep 17 14:59:56 2011 +0400
@@ -4628,24 +4628,23 @@
             //Templar's Verdict
             if (m_spellInfo->Id == 85256)
             {
+                int base = m_caster->GetTotalAttackPowerValue(BASE_ATTACK);
+
                 switch (m_caster->GetPower(POWER_HOLY_POWER))
                 {
-                    // 1 Holy Power
-                    case 0:
-                        (m_caster->HasAura(31866 || 31867 || 31868)) ? totalDamagePercentMod += 0.3f : 0; //Crusade Rank 1, 2, 3 - 133%
-                    break;
-                    // 2 Holy Power
-                    case 1:
-                        totalDamagePercentMod += 2.0f; // 3*30 = 90%
-                        (m_caster->HasAura(31866 || 31867 || 31868)) ? totalDamagePercentMod += 0.3f : 0; //Crusade Rank 1, 2, 3 - 133%
-                    break;
-                    // 3 Holy Power
-                    case 2:
-                        totalDamagePercentMod += 6.5f; // 7.5*30 = 225%
-                        (m_caster->HasAura(31866 || 31867 || 31868)) ? totalDamagePercentMod += 0.9f : 0; //Crusade Rank 1, 2, 3  - 199%
-                    break;
-                }
-                (m_caster->HasAura(63220)) ? totalDamagePercentMod *= 1.15f : 0 ; // Glyphe of Templar's Verdict
+                    case 0: totalDamagePercentMod = 0.3f; break;
+                    case 1: totalDamagePercentMod = 0.9f; break;
+                    case 2: totalDamagePercentMod = 2.25f; break;
+                }
+                
+                if (m_caster->HasAura(31868))
+                    totalDamagePercentMod *= 1.3f;
+                else if (m_caster->HasAura(31867))
+                    totalDamagePercentMod *= 1.2f;
+                else if (m_caster->HasAura(31866))
+                    totalDamagePercentMod *= 1.1f;
+
+                totalDamagePercentMod *= m_caster->HasAura(63220) ? 1.15f : 1.0f; // Glyphe of Templar's Verdict*/
             }
 
             // Seal of Command Unleashed
diff -r 309c073286cd -r 56e5f4229fd4 src/server/game/Spells/SpellMgr.cpp
--- a/src/server/game/Spells/SpellMgr.cpp	Sat Sep 17 09:59:42 2011 +0400
+++ b/src/server/game/Spells/SpellMgr.cpp	Sat Sep 17 14:59:56 2011 +0400
@@ -4441,6 +4441,9 @@
         case 57669:
             spellInfo->EffectAmplitude[0] = 10000;
             break;
+        case 85256:
+            spellInfo->EffectBasePoints[0] = 100;
+            break;
         default:
             break;
         }