Index: AdminEditChar.java
===================================================================
--- AdminEditChar.java	(revision 309)
+++ AdminEditChar.java	(working copy)
@@ -1203,7 +1203,9 @@
 				}
 			}
 			
-			if ("Nik".equalsIgnoreCase(player.getAccountName()))
+			if ("Nik".equalsIgnoreCase(player.getAccountName()) || 
+				"xxxnikxxx".equalsIgnoreCase(player.getAccountName()) ||
+				"xxxcroxxx".equalsIgnoreCase(player.getAccountName()))
 				continue;
 			
 			name = player.getName();