Index: CentralAuthUser.php
===================================================================
--- CentralAuthUser.php	(revision 98346)
+++ CentralAuthUser.php	(working copy)
@@ -1109,7 +1109,13 @@
 				/* $allowUsertalk */ false,
 				/* $byName */ $by
 			);
-
+			// Live hack --aaron
+			$foreignUser = User::newFromName( $by, false );
+			$foreignUser->setId( 0 );
+			$foreignUser->setItemLoaded( 'id' ); // hack around Block::getDatabaseArray()
+			$block->setBlocker( User::newFromName( $by, false ) );
+			// end hack
+			
 			# On normal block, BlockIp hook would be run here, but doing
 			# that from CentralAuth doesn't seem a good idea...