--- Components/Boiler/boiler.cpp
+++ Components/Boiler/boiler.cpp
@@ -34,8 +34,8 @@
                       name, AITP_WMVEC);
     //默认锅炉特殊参数    
     TCPK_QSTRINGLIST *vard1 = new TCPK_QSTRINGLIST;
-    *(vard1->m_ptr)<<"压力损失系数：P2=P1*(1-DP)"  //0
-                   <<"压力损失绝度值：P2=P1-DP";   //1
+    *(vard1->m_ptr)<<"压力损失系数"  //0
+                   <<"压力损失绝度值";   //1
     vard1->m_ptrelct = 0;
     *((TCPK_DOUBLE *)m_asiComAsyncInfo.m_vptrSpecialVars[BLSVID_PressureDropValue])->m_ptr = 0.123;