--- a/Engine/Plugins/Runtime/Steam/SteamVR/Source/SteamVRController/Private/SteamVRController.cpp
+++ b/Engine/Plugins/Runtime/Steam/SteamVR/Source/SteamVRController/Private/SteamVRController.cpp
@@ -128,3 +128,4 @@ public:
                        System,
-                       ApplicationMenu,
+                       ApplicationMenu, // Steam menu button, Oculus Touch B/Y buttons
+                       OculusTouchAX, // Oculus Touch controller A/X buttons
                        TouchPadPress,
@@ -176,2 +177,3 @@ public:
                Buttons[ (int32)EControllerHand::Left ][ ESteamVRControllerButton::ApplicationMenu ] = FGamepadKeyNames::MotionController_Left_Shoulder;
+               Buttons[ (int32)EControllerHand::Left ][ ESteamVRControllerButton::OculusTouchAX ] = FGamepadKeyNames::MotionController_Left_FaceButton5;
                Buttons[ (int32)EControllerHand::Left ][ ESteamVRControllerButton::TouchPadPress ] = FGamepadKeyNames::MotionController_Left_Thumbstick;
@@ -183,2 +185,3 @@ public:
                Buttons[ (int32)EControllerHand::Right ][ ESteamVRControllerButton::ApplicationMenu ] = FGamepadKeyNames::MotionController_Right_Shoulder;
+               Buttons[ (int32)EControllerHand::Left  ][ ESteamVRControllerButton::OculusTouchAX ] = FGamepadKeyNames::MotionController_Right_FaceButton5;
                Buttons[ (int32)EControllerHand::Right ][ ESteamVRControllerButton::TouchPadPress ] = FGamepadKeyNames::MotionController_Right_Thumbstick;
@@ -193,2 +196,3 @@ public:
                Buttons[ (int32)EControllerHand::Pad ][ ESteamVRControllerButton::ApplicationMenu ] = SteamVRControllerKeyNames::GenericMenu;
+               Buttons[ (int32)EControllerHand::Pad ][ ESteamVRControllerButton::OculusTouchAX ] = FGamepadKeyNames::Invalid;
                Buttons[ (int32)EControllerHand::Pad ][ ESteamVRControllerButton::TouchPadPress ] = SteamVRControllerKeyNames::GenericTouchpad;
@@ -204,2 +208,3 @@ public:
                Buttons[ (int32)EControllerHand::ExternalCamera ][ ESteamVRControllerButton::ApplicationMenu ] = FGamepadKeyNames::Invalid;
+               Buttons[ (int32)EControllerHand::ExternalCamera ][ ESteamVRControllerButton::OculusTouchAX ] = FGamepadKeyNames::Invalid;
                Buttons[ (int32)EControllerHand::ExternalCamera ][ ESteamVRControllerButton::TouchPadPress ] = FGamepadKeyNames::Invalid;
@@ -215,2 +220,3 @@ public:
                Buttons[(int32)EControllerHand::Gun][ESteamVRControllerButton::ApplicationMenu] = SteamVRControllerKeyNames::GenericMenu;
+               Buttons[(int32)EControllerHand::Gun][ESteamVRControllerButton::OculusTouchAX] = FGamepadKeyNames::Invalid;
                Buttons[(int32)EControllerHand::Gun][ESteamVRControllerButton::TouchPadPress] = FGamepadKeyNames::Invalid;
@@ -228,2 +234,3 @@ public:
                        Buttons[SpecialIndex][ESteamVRControllerButton::ApplicationMenu] = SteamVRControllerKeyNames::GenericMenu;
+                       Buttons[SpecialIndex][ESteamVRControllerButton::OculusTouchAX] = FGamepadKeyNames::Invalid;
                        Buttons[SpecialIndex][ESteamVRControllerButton::TouchPadPress] = SteamVRControllerKeyNames::GenericTouchpad;
@@ -483,2 +490,3 @@ public:
                                                CurrentStates[ ESteamVRControllerButton::ApplicationMenu ] = !!(VRControllerState.ulButtonPressed & vr::ButtonMaskFromId(vr::k_EButton_ApplicationMenu));
+                                               CurrentStates[ ESteamVRControllerButton::OculusTouchAX ] = !!(VRControllerState.ulButtonPressed & vr::ButtonMaskFromId(vr::k_EButton_A));
                                                CurrentStates[ ESteamVRControllerButton::TouchPadPress ] = !!(VRControllerState.ulButtonPressed & vr::ButtonMaskFromId(vr::k_EButton_SteamVR_Touchpad));