diff --git a/dep/StormLib/CMakeLists.txt b/dep/StormLib/CMakeLists.txt
index 7096c8f8729..6ecf4b85140 100644
--- a/dep/StormLib/CMakeLists.txt
+++ b/dep/StormLib/CMakeLists.txt
@@ -261,7 +261,8 @@ endif()

 add_library(storm STATIC ${SRC_FILES} ${SRC_ADDITIONAL_FILES})
 target_link_libraries(storm ${LINK_LIBS})
-
+  PRIVATE
+    trinity-dependency-interface
 if(UNIX)
     set_target_properties(storm PROPERTIES SOVERSION 0)
 endif()
* Unmerged path dep/libmpq/CMakeLists.txt
diff --git a/dep/zmqpp/CMakeLists.txt b/dep/zmqpp/CMakeLists.txt
index 381d40b42d0..0958c54e8cc 100644
--- a/dep/zmqpp/CMakeLists.txt
+++ b/dep/zmqpp/CMakeLists.txt
@@ -17,6 +17,8 @@ add_library(zmqpp STATIC
 )

 target_link_libraries(zmqpp
+  PRIVATE
+    trinity-dependency-interface
   PUBLIC
     zmq)