diff --git a/src/js/utilJS.tw b/src/js/utilJS.tw
index 86b6620..a7e513c 100644
--- a/src/js/utilJS.tw
+++ b/src/js/utilJS.tw
@@ -378,6 +378,15 @@ window.Categorizer.prototype.cat = function(val, def) {
 	return result;
 };
 
+window.commaNum = function(s) {
+	if(!s) { return 0; }
+	return s.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+}
+
+window.cashFormat = function(s) {
+	if(!s) { s = 0; }
+	return "¤"+commaNum(s);
+}
 
 window.isFloat = function(n){
     return n === +n && n !== (n|0);
diff --git a/src/pregmod/widgets/economyWidgets.tw b/src/pregmod/widgets/economyWidgets.tw
index 9e367e4..dd156eb 100644
--- a/src/pregmod/widgets/economyWidgets.tw
+++ b/src/pregmod/widgets/economyWidgets.tw
@@ -387,24 +387,24 @@
 <<widget "CorpDevBuySell">>
 <<set _textboxMLArg = '_'+$args[1]>>
 | Trade Qty 
-<<textbox `'_Num' + $args[0]` `eval(parse('_Num' + $args[0]))`>><<script>>setTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>: 
+<<textbox `'_Num' + $args[0]` `eval(parse('_Num' + $args[0]))`>><<script>>setTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>: 
 <<link "Buy">>
 	<<if Number.isInteger(Number.parseFloat(eval(parse('_Num' + $args[0])))) == false>> /* Checks if _Numasset is string */
 /* Buy all */
 		<<if State.temporary[$args[1]].toLowerCase() == "all" || "max">> 
 			<<set State.temporary[$args[1]] = Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))>>
 			<<replace `'#'+$args[0]+'AssetsPurchasing'`>> 
-				Bought <<print Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))>> units for ¤<<print Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))*500*State.variables[$args[0]+"AssetPrice"]>>
+				Bought <<print commaNum(commaNum(Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))))>> units for <<print commaNum(cashFormat(Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))*500*State.variables[$args[0]+"AssetPrice"]))>>
 			<</replace>>
 			<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 			<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 			<<replace '#cashOnHand'>>
-				¤<<print $corpCash>>
+				<<print commaNum(cashFormat($corpCash))>>
 			<</replace>>
 			<<replace `'#'+$args[0]+'AssetsDisp'`>>
-				¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+				<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<print commaNum(Math.floor(State.variables[$args[0]+"Assets"]/500))>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 			<</replace>>
-			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 		<<else>>
 /* Error if decimal or not "all" or "max" string */
 			<<replace `'#'+$args[0]+'AssetsPurchasing'`>> Please input a round number. <</replace>>
@@ -414,35 +414,35 @@
 		<<set _DesPerc = State.temporary[$args[1]]>>
 		<<set State.temporary[$args[1]] = Math.floor((Math.clamp((Number.parseFloat(State.temporary[$args[1]])/100),0,1)*$corpCash)/(State.variables[$args[0]+"AssetPrice"] * 500))>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>>
-			Purchased <<print State.temporary[$args[1]]>> units (<<print (State.temporary[$args[1]]*(State.variables[$args[0]+"AssetPrice"] * 50000)/$corpCash).toFixed(2)>>% of cash, desired was _DesPerc)
+			Purchased <<print commaNum(commaNum(State.temporary[$args[1]]))>> units (<<print commaNum((State.temporary[$args[1]]*(State.variables[$args[0]+"AssetPrice"] * 50000)/$corpCash).toFixed(2))>>% of cash, desired was _DesPerc)
 		<</replace>>
 		<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 		<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print commaNum(cashFormat($corpCash))>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<print commaNum(Math.floor(State.variables[$args[0]+"Assets"]/500))>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
-		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Successful buy */
 	<<elseif (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"] < $corpCash>>
 		<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 		<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print commaNum(cashFormat($corpCash))>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<print commaNum(Math.floor(State.variables[$args[0]+"Assets"]/500))>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>><</replace>>
 /* Unsuccessful buy but have enough cash for more than 1 unit */
 	<<elseif $corpCash > 500 * State.variables[$args[0]+"AssetPrice"]>> 
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>>
-			The corp can only buy <<print Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))>> more unit<<if Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500)) > 1 >>s<</if>> of $args[0] assets. 
+			The corp can only buy <<print commaNum(Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500)))>> more unit<<if Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500)) > 1 >>s<</if>> of $args[0] assets. 
 		<</replace>>
 		<<set State.temporary[$args[1]] = Math.floor($corpCash/(State.variables[$args[0]+"AssetPrice"] * 500))>>
-		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Unsuccessful buy */
 	<<else>> 
 	<<replace `'#'+$args[0]+'AssetsPurchasing'`>>There are insufficient funds for additional purchases.<</replace>> 
@@ -453,18 +453,18 @@
 /* Sell all */
 		<<if State.temporary[$args[1]].toLowerCase() == "all" || "max">>
 			<<replace `'#'+$args[0]+'AssetsPurchasing'`>>
-				Sold <<print Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)>> units for ¤<<print Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)*500*State.variables[$args[0]+"AssetPrice"]>>
+				Sold <<print commaNum(commaNum(Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)))>> units for <<print commaNum(cashFormat(Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)*500*State.variables[$args[0]+"AssetPrice"]))>>
 			<</replace>>
 			<<set State.temporary[$args[1]] = Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)>>
 			<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 			<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 			<<replace '#cashOnHand'>>
-				¤<<print $corpCash>>
+				<<print commaNum(cashFormat($corpCash))>>
 			<</replace>>
 			<<replace `'#'+$args[0]+'AssetsDisp'`>>
-				¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+				<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print commaNum(Math.ceil(State.variables[$args[0]+"Assets"]/500))>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 			<</replace>>
-			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Error if decimal or not "all" or "max" string */
 		<<else>>
 			<<replace `'#'+$args[0]+'AssetsPurchasing'`>> Please input a round number. <</replace>>
@@ -474,38 +474,38 @@
 		<<set _DesPerc = State.temporary[$args[1]]>>
 		<<set State.temporary[$args[1]] = Math.ceil((State.variables[$args[0]+"Assets"]-500)/500*(Math.clamp((Number.parseFloat(State.temporary[$args[1]])/100),0,1)))>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>>
-			Sold <<print State.temporary[$args[1]]>> units (<<print (State.temporary[$args[1]]*100/((State.variables[$args[0]+"Assets"]-500)/500)).toFixed(2)>>% of $args[0] assets, desired was _DesPerc)
+			Sold <<print commaNum(commaNum(State.temporary[$args[1]]))>> units (<<print commaNum((State.temporary[$args[1]]*100/((State.variables[$args[0]+"Assets"]-500)/500)).toFixed(2))>>% of $args[0] assets, desired was _DesPerc)
 		<</replace>>
 		<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 		<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print commaNum(cashFormat($corpCash))>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print commaNum(Math.ceil(State.variables[$args[0]+"Assets"]/500))>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
-		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Successful sell */
 	<<elseif State.variables[$args[0]+"Assets"] > 500 && eval(parse('_Num' + $args[0])) * 500 < State.variables[$args[0]+"Assets"]>>
 		<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 		<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print commaNum(cashFormat($corpCash))>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if (State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print commaNum(cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]))>> (<<if (State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print commaNum(Math.ceil(State.variables[$args[0]+"Assets"]/500))>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>><</replace>>
 /* Unsuccessful sell but have enough assets for more than 1 unit */
 	<<elseif State.variables[$args[0]+"Assets"] > 500>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>>
-			The corp can only sell <<print Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)>> more unit<<if Math.ceil((State.variables[$args[0]+"Assets"]-500)/500) > 1 >>s<</if>> of $args[0] assets. 
+			The corp can only sell <<print commaNum(Math.ceil((State.variables[$args[0]+"Assets"]-500)/500))>> more unit<<if Math.ceil((State.variables[$args[0]+"Assets"]-500)/500) > 1 >>s<</if>> of $args[0] assets. 
 		<</replace>>
 		<<set State.temporary[$args[1]] = Math.ceil((State.variables[$args[0]+"Assets"]-500)/500)>>
-		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 5);<</script>>
+		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Unsuccessful sell */
 	<<else>>
 	<<replace `'#'+$args[0]+'AssetsPurchasing'`>>The corp cannot sell any more $args[0] assets.<</replace>>
 	<</if>>
 <</link>>
-<</widget>>
\ No newline at end of file
+<</widget>>
diff --git a/src/uncategorized/SFMBarracks.tw b/src/uncategorized/SFMBarracks.tw
index dfc4552..c385f7f 100644
--- a/src/uncategorized/SFMBarracks.tw
+++ b/src/uncategorized/SFMBarracks.tw
@@ -50,12 +50,12 @@
 		<<set $CashGift = (Math.ceil((Math.abs($cash)*0.05)*(Math.max(0.99,$SFAO))))*$Env>>
 			<<if random(1,100) > 50>>
 				<<if random(1,100) > 50>>
-					_Name nods. "Sure boss," she says, "we had a bit of a haul this week. One of my sergeants convinced a woman to tell us where she had hidden her shit. Cut her up pretty bad, but she told us. Bunch of nice jewelry, I kept a nice piece for myself." She picks up a tablet on the table, tapping a few commands on it. "@@.green;There's your cut,@@ ¤<<print $CashGift>>."
+					_Name nods. "Sure boss," she says, "we had a bit of a haul this week. One of my sergeants convinced a woman to tell us where she had hidden her shit. Cut her up pretty bad, but she told us. Bunch of nice jewelry, I kept a nice piece for myself." She picks up a tablet on the table, tapping a few commands on it. "@@.green;There's your cut,@@ <<print cashFormat($CashGift)>>."
 				<<else>>
-					_Name smiles widely. "Sure boss," she says, "we pulled in some good shit this week. One of the boys found a real nice family hiding in a basement. 18-year old triplets. Brought in a good bit of cash." She picks up a tablet on the table, tapping a few commands on it. "@@.green;There's your cut,@@ ¤<<print $CashGift>>."
+					_Name smiles widely. "Sure boss," she says, "we pulled in some good shit this week. One of the boys found a real nice family hiding in a basement. 18-year old triplets. Brought in a good bit of cash." She picks up a tablet on the table, tapping a few commands on it. "@@.green;There's your cut,@@ <<print cashFormat($CashGift)>>."
 				<</if>>
 			<<else>>
-				_Name picks up a tablet. "Sure boss," she says, "we had a nice score this week. Looters fucked up and left a bunch of nice shit behind." She taps a few commands on the tablet. "@@.green;There's your cut,@@ ¤<<print $CashGift>>."
+				_Name picks up a tablet. "Sure boss," she says, "we had a nice score this week. Looters fucked up and left a bunch of nice shit behind." She taps a few commands on the tablet. "@@.green;There's your cut,@@ <<print cashFormat($CashGift)>>."
 			<</if>>
 			<<set $securityForceGiftToken = 1>>
 			<<set $cash += $CashGift>>
diff --git a/src/uncategorized/corporationDevelopments.tw b/src/uncategorized/corporationDevelopments.tw
index 2df4bab..cc830b7 100644
--- a/src/uncategorized/corporationDevelopments.tw
+++ b/src/uncategorized/corporationDevelopments.tw
@@ -18,7 +18,7 @@
 <<set $corpValue = ($generalAssets*$generalAssetPrice)+($slaveAssets*$slaveAssetPrice)+($entrapmentAssets*$entrapmentAssetPrice)+($captureAssets*$captureAssetPrice)+($trainingAssets*$trainingAssetPrice)+($surgicalAssets*$surgicalAssetPrice)+($drugAssets*$drugAssetPrice)>>
 <<set $corpProfit = Math.trunc($corpValue*(random(7,15)/1000))>>
 <<set $corpCash += $corpProfit>>
-Your corporation made a profit of ¤$corpProfit last week.
+Your corporation made a profit of <<print cashFormat($corpProfit)>> last week.
 <<if $mercenariesHelpCorp > 0>>
 	The $mercenariesTitle assist it with difficult enslavement targets. Otherwise, it
 	<<set $slaveAssets += $mercenaries*1000>>
@@ -47,7 +47,7 @@ Your corporation made a profit of ¤$corpProfit last week.
 	<<set $slaveAssets += Math.trunc(($captureAssets+$entrapmentAssets)*0.01)>>
 	<<set $corpPeopleEnslaved += Math.trunc(($captureAssets+$entrapmentAssets)*0.0001)>>
 <</if>>
-in total, the corporation has enslaved $corpPeopleEnslaved people.
+in total, the corporation has enslaved <<print commaNum($corpPeopleEnslaved)>> people.
 <<if $slaveAssets < ($trainingAssets+$surgicalAssets+$drugAssets)>>
 	The corporation has enough training and medical assets to rapidly improve its human holdings.
 	<<set $slaveAssets = Math.trunc($slaveAssets*1.1)>>
@@ -74,27 +74,27 @@ in total, the corporation has enslaved $corpPeopleEnslaved people.
 <</if>>
 <</if>>
 <<if $corpProfit*0.6 > ($personalShares+$publicShares)*3>>
-	Since it is incredibly profitable, a massive dividend of ¤<<print Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares))>> was paid out to stockholders; you received @@.yellowgreen;¤<<print Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares))*$personalShares>>.@@
+	Since it is incredibly profitable, a massive dividend of <<print cashFormat(Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares)))>> was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat(Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares))*$personalShares)>>.@@
 	<<set $cash += Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares))*$personalShares>>
 	<<set $corpCash -= Math.trunc(($corpProfit*0.6)/($personalShares+$publicShares))*$personalShares>>
 <<elseif $corpProfit > ($personalShares+$publicShares)*2.4>>
-	Since it is extremely profitable, a huge dividend of ¤2/share was paid out to stockholders; you received @@.yellowgreen;¤<<print $personalShares*2>>.@@
+	Since it is extremely profitable, a huge dividend of ¤2/share was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat($personalShares*2)>>.@@
 	<<set $cash += $personalShares*2>>
 	<<set $corpCash -= ($personalShares+$publicShares)*2>>
 <<elseif $corpProfit > ($personalShares+$publicShares)*1.5>>
-	Since it is highly profitable, a large dividend of ¤1/share was paid out to stockholders; you received @@.yellowgreen;¤$personalShares.@@
+	Since it is highly profitable, a large dividend of ¤1/share was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat($personalShares)>>.@@
 	<<set $cash += $personalShares>>
 	<<set $corpCash -= ($personalShares+$publicShares)>>
 <<elseif $corpProfit > ($personalShares+$publicShares)*0.8>>
-	Since it is very profitable, a moderate dividend of ¤0.5/share was paid out to stockholders; you received @@.yellowgreen;¤<<print $personalShares*0.5>>.@@
+	Since it is very profitable, a moderate dividend of ¤0.5/share was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat($personalShares*0.5)>>.@@
 	<<set $cash += Math.trunc($personalShares*0.5)>>
 	<<set $corpCash -= Math.trunc(($personalShares+$publicShares)*0.5)>>
 <<elseif $corpProfit > ($personalShares+$publicShares)*0.4>>
-	Since it is moderately profitable, a reasonable dividend of ¤0.2/share was paid out to stockholders; you received @@.yellowgreen;¤<<print $personalShares*0.2>>.@@
+	Since it is moderately profitable, a reasonable dividend of ¤0.2/share was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat($personalShares*0.2)>>.@@
 	<<set $cash += Math.trunc($personalShares*0.2)>>
 	<<set $corpCash -= Math.trunc(($personalShares+$publicShares)*0.2)>>
 <<elseif $corpProfit > ($personalShares+$publicShares)*0.2>>
-	Since it is acceptably profitable, a small dividend of ¤0.1/share was paid out to stockholders; you received @@.yellowgreen;¤<<print $personalShares*0.1>>.@@
+	Since it is acceptably profitable, a small dividend of ¤0.1/share was paid out to stockholders; you received @@.yellowgreen;<<print cashFormat($personalShares*0.1)>>.@@
 	<<set $cash += Math.trunc($personalShares*0.1)>>
 	<<set $corpCash -= Math.trunc(($personalShares+$publicShares)*0.1)>>
 <</if>>
@@ -118,19 +118,19 @@ in total, the corporation has enslaved $corpPeopleEnslaved people.
 	<<set $sharePrice = (Math.trunc($sharePrice*random(85,random(90,95))))/100>>
 <</if>>
 
-Shares in your corporation are trading at ¤$sharePrice:
+Shares in your corporation are trading at <<print cashFormat($sharePrice)>>:
 
 <<if $oldSharePrice == $sharePrice>>
 	no different from last week.
 <<elseif $oldSharePrice > $sharePrice>>
-	down <<print Math.trunc(($oldSharePrice-$sharePrice)*100)/100>> since last week.
+	down <<print cashFormat(Math.trunc(($oldSharePrice-$sharePrice)*100)/100)>> since last week.
 <<else>>
-	up <<print Math.trunc(($sharePrice - $oldSharePrice)*100)/100>> since last week.
+	up <<print cashFormat(Math.trunc(($sharePrice - $oldSharePrice)*100)/100)>> since last week.
 <</if>>
 
 <<set $oldSharePrice = Math.trunc($sharePrice)>>
 
-You hold $personalShares shares personally while $publicShares are publicly held.
+You hold <<print commaNum($personalShares)>> shares personally while <<print commaNum($publicShares)>> are publicly held.
 <span id="CorpAction">
 <br>Purchase shares from corporation:
 <<if $cash > 1000*$sharePrice>>
@@ -189,7 +189,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <br>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <span id="purchasing"></span>
-<span id="cashOnHand">¤$corpCash </span>in cash on hand.
+<span id="cashOnHand"><<print cashFormat($corpCash)>></span> in cash on hand.
 <span id="update"></span>
 
 <br>
@@ -216,7 +216,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="genAssets">¤<<print $generalAssetPrice*$generalAssets>></span> in general assets.
+	<span id="genAssets"><<print cashFormat($generalAssetPrice*$generalAssets)>></span> in general assets.
 	General asset prices are
 	<<if $generalAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $generalAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $generalAssetPrice == 10>>average<<elseif $generalAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -224,10 +224,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $generalAssets += 500>>
 			<<set $corpCash -= 500*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -242,10 +242,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $generalAssets += (Math.trunc($corpCash/($generalAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($generalAssetPrice*500)))*($generalAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 		<<else>>
 		<<replace "#purchasing">>
@@ -258,10 +258,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $generalAssets -= 500>>
 			<<set $corpCash += 500*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -278,10 +278,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $generalAssets = Math.trunc($generalAssets/2)>>
 			<<set $corpCash += $generalAssets*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -292,14 +292,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="generalAssetsDisp">¤<<print $generalAssetPrice*$generalAssets>> (<<print $generalAssets/500>> unit<<if $generalAssets/500 > 1>>s<</if>>)</span> in general assets.
+	<span id="generalAssetsDisp"><<print cashFormat($generalAssetPrice*$generalAssets)>> (<<print commaNum($generalAssets/500)>> unit<<if $generalAssets/500 > 1>>s<</if>>)</span> in general assets.
 	Prices are 
 	<<if $generalAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $generalAssetPrice*500>>ea@@.
+		@@.yellowgreen;<<print cashFormat($generalAssetPrice*500)>>ea@@.
 	<<elseif $generalAssetPrice < 10>>
-		@@.gold;¤<<print $generalAssetPrice*500>>ea@@.
+		@@.gold;<<print cashFormat($generalAssetPrice*500)>>ea@@.
 	<<else>>
-		¤<<print $generalAssetPrice*500>>ea.
+		<<print cashFormat($generalAssetPrice*500)>>ea.
 	<</if>>
 	<span id="controlsgeneral"><<CorpDevBuySell "general" "Numgeneral">></span><span id='generalAssetsPurchasing'></span>
 <</if>>
@@ -308,7 +308,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <<set $slaveAssetPrice = Math.trunc(10*$slaveCostFactor)>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="slavAssets">¤<<print $slaveAssetPrice*$slaveAssets>></span> in slaves.
+	<span id="slavAssets"><<print cashFormat($slaveAssetPrice*$slaveAssets)>></span> in slaves.
 	Slave prices are
 	<<if $slaveAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $slaveAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $slaveAssetPrice == 10>>average<<elseif $slaveAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -317,10 +317,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $corpPeopleEnslaved += Math.trunc((500*$slaveAssetPrice)/1000)>>
 			<<set $corpCash -= 500*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -336,10 +336,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $corpPeopleEnslaved += Math.trunc(($corpCash/($slaveAssetPrice*2))/1000)>>
 			<<set $corpCash -= (Math.trunc($corpCash/($slaveAssetPrice*500)))*($slaveAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -352,10 +352,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $slaveAssets -= 500>>
 			<<set $corpCash += 500*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -370,10 +370,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $slaveAssets = Math.trunc($slaveAssets/2)>>
 			<<set $corpCash += $slaveAssets*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -384,14 +384,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="slaveAssetsDisp">¤<<print $slaveAssetPrice*$slaveAssets>> (<<print Math.floor($slaveAssets/500)>> unit<<if Math.floor($slaveAssets/500) > 1>>s<</if>>)</span> in slaves.
+	<span id="slaveAssetsDisp"><<print cashFormat($slaveAssetPrice*$slaveAssets)>> (<<print commaNum(Math.floor($slaveAssets/500))>> unit<<if Math.floor($slaveAssets/500) > 1>>s<</if>>)</span> in slaves.
 	Prices are 
 	<<if $slaveAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $slaveAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($slaveAssetPrice*500)>>ea@@
 	<<elseif $slaveAssetPrice < 10>>
-		@@.gold;¤<<print $slaveAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($slaveAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $slaveAssetPrice*500>>ea
+		<<print cashFormat($slaveAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsslave"><<CorpDevBuySell "slave" "Numslave">></span><span id='slaveAssetsPurchasing'></span>
 <</if>>
@@ -422,7 +422,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="trapAssets">¤<<print $entrapmentAssetPrice*$entrapmentAssets>></span> in legal enslavement assets.
+	<span id="trapAssets"><<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>></span> in legal enslavement assets.
 	Legal enslavement asset prices are
 	<<if $entrapmentAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $entrapmentAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $entrapmentAssetPrice == 10>>average<<elseif $entrapmentAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -430,10 +430,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $entrapmentAssets += 500>>
 			<<set $corpCash -= 500*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -448,10 +448,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $entrapmentAssets += (Math.trunc($corpCash/($entrapmentAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($entrapmentAssetPrice*500)))*($entrapmentAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -464,10 +464,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $entrapmentAssets -= 500>>
 			<<set $corpCash += 500*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -482,10 +482,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $entrapmentAssets = Math.trunc($entrapmentAssets/2)>>
 			<<set $corpCash += $entrapmentAssets*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -496,14 +496,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="entrapmentAssetsDisp">¤<<print $entrapmentAssetPrice*$entrapmentAssets>> (<<print Math.floor($entrapmentAssets/500)>> unit<<if Math.floor($entrapmentAssets/500) > 1>>s<</if>>)</span> in legal enslavement assets.
+	<span id="entrapmentAssetsDisp"><<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>> (<<print commaNum(Math.floor($entrapmentAssets/500))>> unit<<if Math.floor($entrapmentAssets/500) > 1>>s<</if>>)</span> in legal enslavement assets.
 	Prices are 
 	<<if $entrapmentAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $entrapmentAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($entrapmentAssetPrice*500)>>ea@@
 	<<elseif $entrapmentAssetPrice < 10>>
-		@@.gold;¤<<print $entrapmentAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($entrapmentAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $entrapmentAssetPrice*500>>ea
+		<<print cashFormat($entrapmentAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsentrapment"><<CorpDevBuySell "entrapment" "Numentrapment">></span><span id='entrapmentAssetsPurchasing'></span>
 <</if>>
@@ -534,7 +534,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="capAssets">¤<<print $captureAssetPrice*$captureAssets>></span> in conflict zone enslavement assets.
+	<span id="capAssets"><<print cashFormat($captureAssetPrice*$captureAssets)>></span> in conflict zone enslavement assets.
 	Extralegal enslavement asset prices are
 	<<if $captureAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $captureAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $captureAssetPrice == 10>>average<<elseif $captureAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -542,10 +542,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $captureAssets += 500>>
 			<<set $corpCash -= 500*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -560,10 +560,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $captureAssets += (Math.trunc($corpCash/($captureAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($captureAssetPrice*500)))*($captureAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -576,10 +576,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $captureAssets -= 500>>
 			<<set $corpCash += 500*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -594,10 +594,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $captureAssets = Math.trunc($captureAssets/2)>>
 			<<set $corpCash += $captureAssets*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -608,14 +608,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="captureAssetsDisp">¤<<print $captureAssetPrice*$captureAssets>> (<<print Math.floor($captureAssets/500)>> unit<<if Math.floor($captureAssets/500) > 1>>s<</if>>)</span> in conflict zone enslavement assets.
+	<span id="captureAssetsDisp"><<print cashFormat($captureAssetPrice*$captureAssets)>> (<<print commaNum(Math.floor($captureAssets/500))>> unit<<if Math.floor($captureAssets/500) > 1>>s<</if>>)</span> in conflict zone enslavement assets.
 	Prices are 
 	<<if $captureAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $captureAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($captureAssetPrice*500)>>ea@@
 	<<elseif $captureAssetPrice < 10>>
-		@@.gold;¤<<print $captureAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($captureAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $captureAssetPrice*500>>ea
+		<<print cashFormat($captureAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlscapture"><<CorpDevBuySell "capture" "Numcapture">></span><span id='captureAssetsPurchasing'></span>
 <</if>>
@@ -646,7 +646,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="trainAssets">¤<<print $trainingAssetPrice*$trainingAssets>></span> in slave training assets.
+	<span id="trainAssets"><<print cashFormat($trainingAssetPrice*$trainingAssets)>></span> in slave training assets.
 	Slave training asset prices are
 	<<if $trainingAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $trainingAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $trainingAssetPrice == 10>>average<<elseif $trainingAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -654,10 +654,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $trainingAssets += 500>>
 			<<set $corpCash -= 500*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -672,10 +672,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $trainingAssets += (Math.trunc($corpCash/($trainingAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($trainingAssetPrice*500)))*($trainingAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -688,10 +688,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $trainingAssets -= 500>>
 			<<set $corpCash += 500*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -706,10 +706,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $trainingAssets = Math.trunc($trainingAssets/2)>>
 			<<set $corpCash += $trainingAssets*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -720,14 +720,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="trainingAssetsDisp">¤<<print $trainingAssetPrice*$trainingAssets>> (<<print Math.floor($trainingAssets/500)>> unit<<if Math.floor($trainingAssets/500) > 1>>s<</if>>)</span> of slave training assets.
+	<span id="trainingAssetsDisp"><<print cashFormat($trainingAssetPrice*$trainingAssets)>> (<<print commaNum(Math.floor($trainingAssets/500))>> unit<<if Math.floor($trainingAssets/500) > 1>>s<</if>>)</span> of slave training assets.
 	Prices are 
 	<<if $trainingAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $trainingAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($trainingAssetPrice*500)>>ea@@
 	<<elseif $trainingAssetPrice < 10>>
-		@@.gold;¤<<print $trainingAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($trainingAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $trainingAssetPrice*500>>ea
+		<<print cashFormat($trainingAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlstraining"><<CorpDevBuySell "training" "Numtraining">></span><span id='trainingAssetsPurchasing'></span>
 <</if>>
@@ -758,7 +758,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="surgAssets">¤<<print $surgicalAssetPrice*$surgicalAssets>></span> in surgical assets.
+	<span id="surgAssets"><<print cashFormat($surgicalAssetPrice*$surgicalAssets)>></span> in surgical assets.
 	Surgical asset prices are
 	<<if $surgicalAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $surgicalAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $surgicalAssetPrice == 10>>average<<elseif $surgicalAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -766,10 +766,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $surgicalAssets += 500>>
 			<<set $corpCash -= 500*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -784,10 +784,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $surgicalAssets += (Math.trunc($corpCash/($surgicalAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($surgicalAssetPrice*500)))*($surgicalAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -800,10 +800,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $surgicalAssets -= 500>>
 			<<set $corpCash += 500*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -818,10 +818,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $surgicalAssets = Math.trunc($surgicalAssets/2)>>
 			<<set $corpCash += $surgicalAssets*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -832,14 +832,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="surgicalAssetsDisp">¤<<print $surgicalAssetPrice*$surgicalAssets>> (<<print Math.floor($surgicalAssets/500)>> unit<<if Math.floor($surgicalAssets/500) > 1>>s<</if>>)</span> in surgical assets.
+	<span id="surgicalAssetsDisp"><<print cashFormat($surgicalAssetPrice*$surgicalAssets)>> (<<print commaNum(Math.floor($surgicalAssets/500))>> unit<<if Math.floor($surgicalAssets/500) > 1>>s<</if>>)</span> in surgical assets.
 	Prices are 
 	<<if $surgicalAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $surgicalAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($surgicalAssetPrice*500)>>ea@@
 	<<elseif $surgicalAssetPrice < 10>>
-		@@.gold;¤<<print $surgicalAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($surgicalAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $surgicalAssetPrice*500>>ea
+		<<print cashFormat($surgicalAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlssurgical"><<CorpDevBuySell "surgical" "Numsurgical">></span><span id='surgicalAssetsPurchasing'></span>
 <</if>>
@@ -870,7 +870,7 @@ You hold $personalShares shares personally while $publicShares are publicly held
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="drAssets">¤<<print $drugAssetPrice*$drugAssets>></span> in pharmaceutical assets.
+	<span id="drAssets"><<print cashFormat($drugAssetPrice*$drugAssets)>></span> in pharmaceutical assets.
 	Drug asset prices are
 	<<if $drugAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $drugAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $drugAssetPrice == 10>>average<<elseif $drugAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -878,10 +878,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $drugAssets += 500>>
 			<<set $corpCash -= 500*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -896,10 +896,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $drugAssets += (Math.trunc($corpCash/($drugAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($drugAssetPrice*500)))*($drugAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -912,10 +912,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $drugAssets -= 500>>
 			<<set $corpCash += 500*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -930,10 +930,10 @@ You hold $personalShares shares personally while $publicShares are publicly held
 			<<set $drugAssets = Math.trunc($drugAssets/2)>>
 			<<set $corpCash += $drugAssets*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -944,14 +944,14 @@ You hold $personalShares shares personally while $publicShares are publicly held
 		<</if>>
 	<</link>>
 <<else>>
-	<span id="drugAssetsDisp">¤<<print $drugAssetPrice*$drugAssets>> (<<print Math.floor($drugAssets/500)>> unit<<if Math.floor($drugAssets/500) > 1>>s<</if>>)</span> in pharmaceutical assets.
+	<span id="drugAssetsDisp"><<print cashFormat($drugAssetPrice*$drugAssets)>> (<<print commaNum(Math.floor($drugAssets/500))>> unit<<if Math.floor($drugAssets/500) > 1>>s<</if>>)</span> in pharmaceutical assets.
 	Prices are 
 	<<if $drugAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $drugAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($drugAssetPrice*500)>>ea@@
 	<<elseif $drugAssetPrice < 10>>
-		@@.gold;¤<<print $drugAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($drugAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $drugAssetPrice*500>>ea
+		<<print cashFormat($drugAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsdrug"><<CorpDevBuySell "drug" "Numdrug">></span><span id='drugAssetsPurchasing'></span>
 <</if>>
diff --git a/src/uncategorized/storyCaption.tw b/src/uncategorized/storyCaption.tw
index a0c2732..fd7f79c 100644
--- a/src/uncategorized/storyCaption.tw
+++ b/src/uncategorized/storyCaption.tw
@@ -87,16 +87,16 @@
 <<else>>
 	__@@.red;Cash@@__
 <</if>>
-| ''¤''$cash
+| <<print cashFormat($cash)>>
 </span>
 <br>
 <<if _Pass == "Main">>
 	<<set _TCash2 = ($cash-$cashLastWeek)>>
 	<span id="oldcash">
 	<<if _TCash2 < 0>>
-		(@@.red;_TCash2@@
+		(@@.red;<<print cashFormat(_TCash2)>>@@
 	<<else>>
-		(@@.yellowgreen;+_TCash2@@
+		(@@.yellowgreen;+<<print cashFormat(_TCash2)>>@@
 	<</if>>
 	</span>
 	since last week)
@@ -116,9 +116,9 @@
 		<<set _TCash2 = ($cash-$cashLastWeek)>>
 		<<replace "#oldcash">>
 		<<if _TCash2 < 0>>
-			(@@.red;_TCash2@@
+			(@@.red;<<print cashFormat(_TCash2)>>@@
 		<<else>>
-			(@@.yellowgreen;+_TCash2@@
+			(@@.yellowgreen;+<<print cashFormat(_TCash2)>>@@
 		<</if>>
 		<</replace>>
 		since last week)