--- ext4.h      2011-06-30 14:05:45.000000000 +0400
+++ ext4.h.patched      2011-09-30 11:46:19.058797666 +0400
@@ -921,11 +921,12 @@
        __le16  s_reserved_pad;
        __le64  s_kbytes_written;       /* nr of lifetime kilobytes written */
 //Patch by QNAP:Search filename use case insensitive method
-#ifdef QNAP_SEARCH_FILENAME_CASE_INSENSITIVE
+//#ifdef QNAP_SEARCH_FILENAME_CASE_INSENSITIVE
   #ifdef QNAP_ONLINE_RESIZE_INODE
-  __u32   s_reserved[157];        /* Padding to the end of the block */
   __le32  s_feature_qnap_compat;  /* QNAP internal features */
   __le32  s_rsved_gdt_blks_large; /* Use s_rsved_gdt_blks_large to replace s_reserved_gdt_blocks */
+    #ifdef QNAP_SEARCH_FILENAME_CASE_INSENSITIVE
+  __u32   s_reserved[157];        /* Padding to the end of the block */
   #else
   __u32   s_reserved[159];        /* Padding to the end of the block */
   #endif /* #ifdef QNAP_ONLINE_RESIZE_INODE */