===================================================================
--- Current Version
+++ New Version
@@ -1,9 +1,9 @@
 // ==UserScript==
 // @name			Harem Heroes++
 // @namespace		haremheroes.com
 // @description		Adding things here and there in Harem Heroes game.
-// @version			0.05.2
+// @version			0.05.1
 // @match			http://nutaku.haremheroes.com/*
 // @match			https://www.hentaiheroes.com/*
 // @match			https://www.gayharem.com/*
 // @run-at			document-end
@@ -350,10 +350,9 @@
 	// get haremettes stats & display wiki link
 	i = 0;
 	$('#harem_right').children('[girl]').each( function() {
 		// display: wiki link
-		$(this).append('<div class="WikiLink"><a href="http://harem-battle.club/wiki/Harem-Heroes/HH:' + HList[i].Name + '" target="_blank">Wiki</a></div>');
-        console.log(HList[i]);
+		$(this).append('<div class="WikiLink"><a href="http://harem-battle.club/wiki/Harem-Heroes/HH:' + HList[i].Name + '" target="_blank"> her wiki page </a></div>');
 		i++;
 
 		var j = 0,
 			Taffection = 0, // T= Total requirements (right tooltip)
@@ -371,19 +370,19 @@
 		IncHourly += parseInt($(this).find('.salary').text(), 10);
 
 		girl_quests.find('g').each( function() {
 			// prepare affection tooltip
-			var Raffection = EvoReq[j].affection * 417 * i, // R= Required for this star (right tooltip) // 417 price of 1 affection
-				Rmoney =  EvoReq[j].money * i,
+			var Raffection = EvoReq[j].affection * i, // R= Required for this star (right tooltip)
+				Rmoney = EvoReq[j].money * i,
 				Rkobans = EvoReq[j].kobans * i;
 			Taffection += Raffection;
 			Tmoney += Rmoney;
 			Tkobans += Rkobans;
 			j++;
 			AffectionTT += '<b>' + j + '</b><span class="imgStar"></span> : '
-						 + NbCommas(Raffection + Rmoney) + ' <span class="imgMoney"></span> = '
-						 + '' + NbCommas(Raffection) + ' <span class="imgMoney"></span> (aff) + '
-						 + '(' + NbCommas(Rmoney) + ' <span class="imgMoney"></span> (quest) or ' + NbCommas(Rkobans) + ' <span class="imgKobans"></span>)<br/>';
+						 + NbCommas(Raffection) + ' affection, '
+						 + NbCommas(Rmoney) + ' <span class="imgMoney"></span> or '
+						 + NbCommas(Rkobans) + ' <span class="imgKobans"></span><br />';
 
 			// get stats: unlocked/available scenes & prepare scenes link
 			AvailableSc++;
 			ScenesLink += (ScenesLink === '') ? 'hh_scenes=' : ',';
@@ -418,11 +417,11 @@
 		});
 		ScenesLink = '';
 
 		AffectionTT += '<b>Total:</b> '
-					 + NbCommas(Taffection + Tmoney) + ' <span class="imgMoney"></span> = '
-					 + '' + NbCommas(Taffection) + ' <span class="imgMoney"></span> (aff) + '
-					 + '(' + NbCommas(Tmoney) + ' <span class="imgMoney"></span> (quest) or ' + NbCommas(Tkobans) + ' <span class="imgKobans"></span>)';
+					 + NbCommas(Taffection) + ' affection, '
+					 + NbCommas(Tmoney) + ' <span class="imgMoney"></span> or '
+					 + NbCommas(Tkobans) + ' <span class="imgKobans"></span>';
 
 		// display: Affection costs tooltip
 		girl_quests.parent().children('h4').prepend('<span class="CustomTT"></span><div class="AffectionTooltip">' + AffectionTT + '</div>');
 	});
@@ -483,11 +482,11 @@
 					+ '<span class="subTitle">Money incomes:</span>'
 					+ '~' + NbCommas(IncHourly) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png"> per hour'
 					+ '<br />' + NbCommas(IncCollect) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png"> when all collectable'
 					+ '<span class="subTitle">Required to unlock all locked scenes:</span>'
-					+ NbCommas(Saffection + Smoney) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png">: '
-					+ '<br/> - ' + NbCommas(Saffection) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png"> for affections '
-					+ '<br/> - ' + NbCommas(Smoney) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png"> for quests'
+					+ '- ' + NbCommas(Saffection) + ' affection'
+					+ '<br />- ' + NbCommas(Smoney) + ' <img src="http://i.harem-battle.club/images/2017/01/07/0Gsvn.png">'
+						+ ' or ' + NbCommas(Skobans) + ' <img src="http://i.harem-battle.club/images/2016/08/30/gNUo3XdY.png">'
 					+ '<span class="subTitle">My Stocks:</span>'
 					+ MarketStocks
 					+ '</div>';
 
@@ -632,10 +631,10 @@
 
 	sheet.insertRule('#harem_right .AffectionTooltip {'
 						+ 'position: absolute;'
 						+ 'z-index: 99;'
-						+ 'margin: -130px 0 0 -38px;'
-						+ 'width: 390px;'
+						+ 'margin: -130px 0 0 -28px;'
+						+ 'width: 280px;'
 						+ 'height: 127px;'
 						+ 'border: 1px solid rgb(162, 195, 215);'
 						+ 'border-radius: 8px;'
 						+ 'box-shadow: 0px 0px 4px 0px rgba(0,0,0,0.1);'