diff --git a/core/core/GameModel.js b/core/core/GameModel.js
index 964f0f0..05c2c2d 100644
--- a/core/core/GameModel.js
+++ b/core/core/GameModel.js
@@ -214,7 +214,7 @@ export default class GameModel extends PIXI.utils.EventEmitter {
             if (winlines[i].trigger === trigger) return true;
         }
         let winscatters = this.current('winscatters');
-        for (let i = 0; i < winlines.length; ++i) {
+        for (let i = 0; i < winscatters.length; ++i) {
             if (winscatters[i].trigger === trigger) return true;
         }
         return false;