From 8b7325a2ee22ca16897675d51c587b6e729d52c6 Mon Sep 17 00:00:00 2001
From: Christoph Haag <haagch@frickel.club>
Date: Thu, 18 Jan 2018 09:27:04 +0100
Subject: [PATCH] fix amdgpu.ids path and define

---
 amdgpu/meson.build | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/amdgpu/meson.build b/amdgpu/meson.build
index 55ab9d1d..8b045205 100644
--- a/amdgpu/meson.build
+++ b/amdgpu/meson.build
@@ -19,7 +19,7 @@
 # SOFTWARE.
 
 
-datadir_amdgpu = join_paths(get_option('datadir'), 'libdrm', 'amdgpu.ids')
+datadir_amdgpu = join_paths(get_option('prefix'), get_option('datadir'), 'libdrm')
 
 libdrm_amdgpu = shared_library(
   'drm_amdgpu',
@@ -33,7 +33,7 @@ libdrm_amdgpu = shared_library(
   ],
   c_args : [
     warn_c_args,
-    '-DAMDGPU_ASIC_ID_TABLE="@0@"'.format(datadir_amdgpu),
+    '-DAMDGPU_ASIC_ID_TABLE="@0@"'.format(join_paths(datadir_amdgpu, 'amdgpu.ids')),
   ],
   include_directories : [inc_root, inc_drm],
   link_with : libdrm,
-- 
2.16.0