Index: java/com/l2jserver/gameserver/model/L2Skill.java
===================================================================
--- java/com/l2jserver/gameserver/model/L2Skill.java	(revision 128)
+++ java/com/l2jserver/gameserver/model/L2Skill.java	(working copy)
@@ -1875,18 +1875,7 @@
 							// casting and pvp condition not met
 							if (!player.checkPvpSkill(obj, this))
 								continue;
-							
-							if(NexusEvents.isInEvent(player))
-							{
-								if(!NexusEvents.isInEvent(obj))
-									continue;
-								if(this.isOffensive() && !NexusEvents.canAttack(player, obj))
-									continue;
 								
-								if(!this.isOffensive() && !NexusEvents.canSupport(player, obj))
-									continue;
-							}
-								
 							if (!onlyFirst && addSummon(activeChar, obj, radius, isCorpseType))
 								targetList.add(obj.getPet());
 							
@@ -1964,18 +1953,6 @@
 							if (!player.checkPvpSkill(obj, this))
 								continue;
 							
-							if(NexusEvents.isInEvent(player))
-							{
-								if(!NexusEvents.isInEvent(obj))
-									continue;
-								
-								if(this.isOffensive() && !NexusEvents.canAttack(player, obj))
-									continue;
-								
-								if(!this.isOffensive() && !NexusEvents.canSupport(player, obj))
-									continue;
-							}
-							
 							if (!onlyFirst && addSummon(activeChar, obj, radius, isCorpseType))
 								targetList.add(obj.getPet());
 							
@@ -2082,18 +2059,6 @@
 					if (!player.checkPvpSkill(obj, this))
 						continue;
 					
-					if(NexusEvents.isInEvent(player))
-					{
-						if(!NexusEvents.isInEvent(obj))
-							continue;
-						
-						if(this.isOffensive() && !NexusEvents.canAttack(player, obj))
-							continue;
-						
-						if(!this.isOffensive() && !NexusEvents.canSupport(player, obj))
-							continue;
-					}
-					
 					if (!onlyFirst && addSummon(activeChar, obj, radius, false))
 						targetList.add(obj.getPet());
 					
Index: java/com/l2jserver/gameserver/model/actor/instance/L2PcInstance.java
===================================================================
--- java/com/l2jserver/gameserver/model/actor/instance/L2PcInstance.java	(revision 129)
+++ java/com/l2jserver/gameserver/model/actor/instance/L2PcInstance.java	(working copy)
@@ -9446,6 +9446,17 @@
 		if (getEvent() != null && getEvent().checkPvpSkill(this, target, skill))
 			return false;
 		
+		if(NexusEvents.isInEvent(this))
+		{
+			if(!NexusEvents.isInEvent((L2PcInstance) target))
+				return false;
+			if(skill.isOffensive() && !NexusEvents.canAttack(this, (L2Character) target))
+				return false;
+			
+			if(!skill.isOffensive() && !NexusEvents.canSupport(this, (L2Character) target))
+				return false;
+		}
+		
 		if (
 				target != null &&                                           			// target not null and
 				target != this &&                                           			// target is not self and