Spell/Druid: King of the Jungle

diff -r 4b617b39d4e5 -r d41d95f54a82 sql/updates/world/2011_09_17_00_spell_dru_king_of_the_jungle.sql
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/sql/updates/world/2011_09_17_00_spell_dru_king_of_the_jungle.sql	Sat Sep 17 20:49:00 2011 +0400
@@ -0,0 +1,2 @@
+REPLACE INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (5217, 'spell_dru_king_of_the_jungle');
+REPLACE INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES (5229, 'spell_dru_king_of_the_jungle');
diff -r 4b617b39d4e5 -r d41d95f54a82 src/server/scripts/Spells/spell_druid.cpp
--- a/src/server/scripts/Spells/spell_druid.cpp	Sat Sep 17 19:12:50 2011 +0400
+++ b/src/server/scripts/Spells/spell_druid.cpp	Sat Sep 17 20:49:00 2011 +0400
@@ -272,6 +272,64 @@
         }
 };
 
+// 5229 Enrage
+// 5217 Name: Tiger's Fury
+class spell_dru_king_of_the_jungle : public SpellScriptLoader
+{
+    public:
+        spell_dru_king_of_the_jungle() : SpellScriptLoader("spell_dru_king_of_the_jungle") {}
+
+        class spell_dru_king_of_the_jungle_SpellScript : public SpellScript
+        {
+            PrepareSpellScript(spell_dru_king_of_the_jungle_SpellScript);
+
+            void HandleDummy(SpellEffIndex /*effIndex*/)
+            {
+                if (Unit* caster = GetCaster())
+                {
+                    if (caster->getClass() != CLASS_DRUID)
+                        return;
+
+                    if (caster->HasAura(48492) || caster->HasAura(48494) || caster->HasAura(48495))
+                    {
+                        int spellId;
+                        if (caster->HasAura(48495))
+                            spellId = 48495;
+                        else if (caster->HasAura(48494))
+                            spellId = 48494;
+                        else if(caster->HasAura(48492))
+                            spellId = 48492;
+                        SpellEntry const* spell_entry = sSpellStore.LookupEntry(spellId);
+
+                        int bp0 = 0;
+                        int castSpell = 0;
+                        if (caster->GetShapeshiftForm() == FORM_BEAR)
+                        {
+                            bp0 = spell_entry->EffectBasePoints[0];
+                            castSpell = 51185;
+                        }
+                        else if (caster->GetShapeshiftForm() == FORM_CAT)
+                        {
+                            bp0 = spell_entry->EffectBasePoints[1];
+                            castSpell = 51178;
+                        }
+                        caster->CastCustomSpell(caster, castSpell, &bp0, NULL, NULL, true);
+                    }
+                }
+            }
+
+            void Register()
+            {
+                OnEffect += SpellEffectFn(spell_dru_king_of_the_jungle_SpellScript::HandleDummy, EFFECT_0, SPELL_EFFECT_APPLY_AURA);
+            }
+        };
+
+        SpellScript* GetSpellScript() const
+        {
+            return new spell_dru_king_of_the_jungle_SpellScript();
+        }
+};
+
 void AddSC_druid_spell_scripts()
 {
     new spell_dru_savage_defense();
@@ -280,4 +338,5 @@
     new spell_dru_primal_tenacity();
     new spell_dru_berserk();
     new spell_druid_rejuvenation();
+    new spell_dru_king_of_the_jungle();
 }
\ No newline at end of file