diff --git a/PKGBUILD b/PKGBUILD
index 1e43263..06f99e6 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -5,7 +5,7 @@
 # Contributor: Matthias Sobczyk <matthias.sobczyk@googlemail.com>
 # Contributor: Kamil Kaminski <kyle@kkaminsk.com>
 pkgname=readymedia-transcode-git
-pkgver=581.290ef09
+pkgver=734.38e9139
 pkgrel=1
 pkgdesc="A DLNA/UPnP-AV Media server with transcode support"
 arch=('i686' 'x86_64')
@@ -19,17 +19,23 @@ replaces=('minidlna-transcode-hg')
 backup=('etc/minidlna.conf'
         'etc/conf.d/minidlna')
 source=('readymedia-transcode::git+https://bitbucket.org/stativ/readymedia-transcode.git#branch=transcode' \
-        minidlna.rc tmpfiles.conf minidlna.service)
+        minidlna.rc tmpfiles.conf minidlna.service imagemagick.patch)
 md5sums=('SKIP'
          '484bb94e886616afcb3aba00926aeee9'
          'af1b8c0fef9902f797b0324c255da5a1'
-         'e4d6766af90599929155ff15427bfdf5')
+         'e4d6766af90599929155ff15427bfdf5'
+         '005fc1a8d55f5492174feabf09795ed7')
 
 pkgver() {
   cd "$srcdir/readymedia-transcode"
   echo $(git rev-list --count HEAD).$(git rev-parse --short HEAD)
 }
 
+prepare() {
+  cd "$srcdir/readymedia-transcode"
+  patch -Np1 -i "$srcdir/imagemagick.patch"
+}
+
 build() {
   cd "$srcdir/readymedia-transcode"
 
diff --git a/imagemagick.patch b/imagemagick.patch
new file mode 100644
index 0000000..fc7f458
--- /dev/null
+++ b/imagemagick.patch
@@ -0,0 +1,26 @@
+diff --git a/dlnameta.c b/dlnameta.c
+index 9dcbd9d..e14cd9b 100644
+--- a/dlnameta.c
++++ b/dlnameta.c
+@@ -21,7 +21,7 @@
+ 
+ #include <sys/stat.h>
+ #include <unistd.h>
+-#include <wand/MagickWand.h>
++#include <MagickWand/MagickWand.h>
+ 
+ #include "libav.h"
+ 
+diff --git a/metadata.c b/metadata.c
+index faf5a44..cf72027 100644
+--- a/metadata.c
++++ b/metadata.c
+@@ -30,7 +30,7 @@
+ #include <sys/errno.h>
+ #include <fcntl.h>
+ 
+-#include <wand/MagickWand.h>
++#include <MagickWand/MagickWand.h>
+ #include <libexif/exif-loader.h>
+ #include <jpeglib.h>
+ #include <setjmp.h>