diff --git a/libr/anal/anal.c b/libr/anal/anal.c
index c4e38d329..8c9f03c4f 100644
--- a/libr/anal/anal.c
+++ b/libr/anal/anal.c
@@ -83,6 +83,7 @@ R_API RAnal *r_anal_new() {
 #if USE_DICT
        anal->dict_refs = dict_new (100, dict_free);
        anal->dict_xrefs = dict_new (100, dict_free);
+       anal->dict_fcns = dict_new (100, dict_free);
 #endif
        anal->zign_path = strdup ("");
        anal->cb_printf = (PrintfCallback) printf;
diff --git a/libr/anal/fcn.c b/libr/anal/fcn.c
index 51f7e6b77..770551990 100644
--- a/libr/anal/fcn.c
+++ b/libr/anal/fcn.c
@@ -1344,6 +1344,10 @@ R_API int r_anal_fcn_add(RAnal *a, ut64 addr, ut64 size, const char *name, int t
 #if FCN_SDB
        sdb_set (DB, sdb_fmt (0, "fcn.0x%08"PFMT64x, addr), "TODO", 0); // TODO: add more info here
 #endif
+
+#if USE_DICT
+       setfcn (a->dict_fcns, addr, fcn);
+#endif
        return append? r_anal_fcn_insert (a, fcn): true;
 }
 
@@ -1675,7 +1679,22 @@ R_API int r_anal_str_to_fcn(RAnal *a, RAnalFunction *f, const char *sig) {
        return true;
 }
 
+void setfcn(dict *m, ut64 addr, RAnalFunction *fcn) {
+       if (fcn) {
+               int size = 1;
+
+               RAnalBlock *first_bb = r_list_first (fcn->bbs);
+               if (first_bb) {
+                       size = first_bb->size;
+               }
+               dict_set (m, addr, size, fcn);
+       }
+}
+
 R_API RAnalFunction *r_anal_get_fcn_at(RAnal *anal, ut64 addr, int type) {
+#if USE_DICT
+       return dict_getu (anal->dict_fcns, addr);
+#endif
 #if USE_NEW_FCN_STORE
        // TODO: type is ignored here? wtf.. we need more work on fcnstore
        // if (root) return r_listrange_find_root (anal->fcnstore, addr);
diff --git a/libr/core/blaze.c b/libr/core/blaze.c
index cf3ab6b37..8d3be4dfa 100644
--- a/libr/core/blaze.c
+++ b/libr/core/blaze.c
@@ -468,6 +468,8 @@ R_API bool core_anal_bbs(RCore *core, const char* input) {
                        r_stack_free (stack);
                }
        }
+       dict_stats (core->anal->dict_fcns);
+
 
        sdb_free (sdb);
        r_list_free (result);
diff --git a/libr/include/r_anal.h b/libr/include/r_anal.h
index 4e63cd03a..b78a8cfec 100644
--- a/libr/include/r_anal.h
+++ b/libr/include/r_anal.h
@@ -3,7 +3,7 @@
 #ifndef R2_ANAL_H
 #define R2_ANAL_H
 
-#define USE_DICT 0
+#define USE_DICT 1
 
 /* use sdb function storage */
 #define FCN_SDB 1
@@ -636,6 +636,7 @@ typedef struct r_anal_t {
 #if USE_DICT
        dict *dict_refs;
        dict *dict_xrefs;
+       dict *dict_fcns;
 #endif
 
        RSpaces meta_spaces;
(END)