diff -U0 .config{,.old}
--- .config	2015-08-19 17:55:57.002932733 +0600
+++ .config.old	2015-08-19 19:56:14.844950712 +0600
@@ -3 +3 @@
-# Linux/x86 4.1.5-gentoo Kernel Configuration
+# Linux/x86 3.19.3-gentoo Kernel Configuration
@@ -10,0 +11 @@
+CONFIG_GENTOO_LINUX_PORTAGE=y
@@ -55 +55,0 @@
-CONFIG_PGTABLE_LEVELS=4
@@ -111,0 +112 @@
+CONFIG_GENERIC_CLOCKEVENTS_BUILD=y
@@ -141 +141,0 @@
-CONFIG_SRCU=y
@@ -154 +153,0 @@
-# CONFIG_RCU_EXPEDITE_BOOT is not set
@@ -187,0 +187 @@
+# CONFIG_INIT_FALLBACK is not set
@@ -197 +196,0 @@
-CONFIG_MULTIUSER=y
@@ -275 +273,0 @@
-CONFIG_HAVE_ARCH_HUGE_VMAP=y
@@ -279 +276,0 @@
-CONFIG_ARCH_HAS_ELF_RANDOMIZE=y
@@ -347 +343,0 @@
-CONFIG_LOCK_SPIN_ON_OWNER=y
@@ -361 +356,0 @@
-# CONFIG_X86_AMD_PLATFORM_DEVICE is not set
@@ -367,0 +363 @@
+# CONFIG_MEMTEST is not set
@@ -414,0 +411 @@
+CONFIG_X86_UP_APIC_MSI=y
@@ -427,0 +425,2 @@
+# CONFIG_MICROCODE_INTEL_EARLY is not set
+# CONFIG_MICROCODE_AMD_EARLY is not set
@@ -432 +431 @@
-CONFIG_X86_DIRECT_GBPAGES=y
+CONFIG_DIRECT_GBPAGES=y
@@ -478 +476,0 @@
-# CONFIG_ZSMALLOC_STAT is not set
@@ -480 +477,0 @@
-# CONFIG_X86_PMEM_LEGACY is not set
@@ -510 +506,0 @@
-CONFIG_HAVE_LIVEPATCH=y
@@ -532 +527,0 @@
-CONFIG_ACPI_SYSTEM_POWER_STATES_SUPPORT=y
@@ -551 +545,0 @@
-CONFIG_ACPI_HOTPLUG_IOAPIC=y
@@ -629 +622,0 @@
-CONFIG_PCI_BUS_ADDR_T_64BIT=y
@@ -653,0 +647 @@
+CONFIG_ARCH_BINFMT_ELF_RANDOMIZE_PIE=y
@@ -831 +824,0 @@
-# CONFIG_NET_ACT_BPF is not set
@@ -839 +832 @@
-# CONFIG_MPLS is not set
+CONFIG_NET_MPLS_GSO=y
@@ -893 +885,0 @@
-CONFIG_REGMAP=y
@@ -927 +918,0 @@
-# CONFIG_BLK_DEV_PMEM is not set
@@ -1188 +1178,0 @@
-# CONFIG_DM_MQ_DEFAULT is not set
@@ -1204 +1193,0 @@
-# CONFIG_DM_LOG_WRITES is not set
@@ -1212,0 +1202 @@
+# CONFIG_I2O is not set
@@ -1240,0 +1231 @@
+# CONFIG_NET_DSA_MV88E6060 is not set
@@ -1241,0 +1233,5 @@
+# CONFIG_NET_DSA_MV88E6131 is not set
+# CONFIG_NET_DSA_MV88E6123_61_65 is not set
+# CONFIG_NET_DSA_MV88E6171 is not set
+# CONFIG_NET_DSA_MV88E6352 is not set
+# CONFIG_NET_DSA_BCM_SF2 is not set
@@ -1248,0 +1245 @@
+# CONFIG_NET_XGENE is not set
@@ -1251 +1247,0 @@
-# CONFIG_NET_CADENCE is not set
@@ -1369 +1364,0 @@
-# CONFIG_INPUT_E3X0_BUTTON is not set
@@ -1419 +1413,0 @@
-CONFIG_DEVMEM=y
@@ -1444,0 +1439 @@
+# CONFIG_SERIAL_MFD_HSU is not set
@@ -1654 +1648,0 @@
-# CONFIG_SENSORS_NCT7904 is not set
@@ -1728 +1721,0 @@
-# CONFIG_CADENCE_WATCHDOG is not set
@@ -1798 +1790,0 @@
-# CONFIG_MFD_DA9150 is not set
@@ -1810,0 +1803 @@
+# CONFIG_MFD_MAX77686 is not set
@@ -1812 +1804,0 @@
-# CONFIG_MFD_MAX77843 is not set
@@ -1817 +1808,0 @@
-# CONFIG_MFD_MT6397 is not set
@@ -1824 +1814,0 @@
-# CONFIG_MFD_RT5033 is not set
@@ -1831 +1820,0 @@
-# CONFIG_MFD_SKY81452 is not set
@@ -1849,0 +1839 @@
+# CONFIG_MFD_TC3589X is not set
@@ -1929 +1918,0 @@
-# CONFIG_USB_GSPCA_TOUPTEK is not set
@@ -2071,0 +2061 @@
+# CONFIG_FB_VIA is not set
@@ -2108,2 +2097,0 @@
-CONFIG_DUMMY_CONSOLE_COLUMNS=80
-CONFIG_DUMMY_CONSOLE_ROWS=25
@@ -2207 +2194,0 @@
-# CONFIG_SND_SE6X is not set
@@ -2241 +2227,0 @@
-CONFIG_SND_HDA_CORE=y
@@ -2264 +2249,0 @@
-# CONFIG_HID_BETOP_FF is not set
@@ -2274,0 +2260 @@
+# CONFIG_HID_HUION is not set
@@ -2357,0 +2344 @@
+# CONFIG_USB_ISP1760_HCD is not set
@@ -2409 +2395,0 @@
-# CONFIG_USB_ISP1760 is not set
@@ -2520,2 +2505,0 @@
-# CONFIG_RTC_DRV_ABB5ZES3 is not set
-# CONFIG_RTC_DRV_ABX80X is not set
@@ -2557 +2540,0 @@
-# CONFIG_RTC_DRV_DS1685_FAMILY is not set
@@ -2583,0 +2567 @@
+# CONFIG_INTEL_MID_DMAC is not set
@@ -2584,0 +2569 @@
+# CONFIG_DW_DMAC_CORE is not set
@@ -2587 +2571,0 @@
-# CONFIG_HSU_DMA_PCI is not set
@@ -2686 +2669,0 @@
-# CONFIG_EXT4_ENCRYPTION is not set
@@ -2697,2 +2679,0 @@
-# CONFIG_F2FS_FS is not set
-# CONFIG_FS_DAX is not set
@@ -2772 +2752,0 @@
-# CONFIG_PSTORE_PMSG is not set
@@ -2777,0 +2758 @@
+# CONFIG_F2FS_FS is not set
@@ -2851 +2831,0 @@
-# CONFIG_GDB_SCRIPTS is not set
@@ -2887,2 +2866,0 @@
-CONFIG_HAVE_ARCH_KASAN=y
-CONFIG_KASAN_SHADOW_OFFSET=0xdffffc0000000000
@@ -2910 +2887,0 @@
-# CONFIG_DEBUG_TIMEKEEPING is not set
@@ -2939 +2915,0 @@
-# CONFIG_PROVE_RCU is not set
@@ -2975 +2950,0 @@
-# CONFIG_TEST_HEXDUMP is not set
@@ -2986 +2960,0 @@
-# CONFIG_MEMTEST is not set
@@ -3156 +3129,0 @@
-# CONFIG_CRYPTO_USER_API_RNG is not set
@@ -3159 +3131,0 @@
-CONFIG_KVM_COMPAT=y
@@ -3167,2 +3138,0 @@
-# CONFIG_HAVE_ARCH_BITREVERSE is not set
-CONFIG_RATIONAL=y