diff -r -U 3 ff7.orig/controls.js ff7/controls.js
--- ff7.orig/controls.js        2015-12-27 17:30:24.000000000 -0800
+++ ff7/controls.js     2017-12-31 16:52:57.905000798 -0800
@@ -313,7 +313,7 @@
 
                offset.x = radius * Math.sin( phi ) * Math.sin( theta );
 //             offset.y = radius * Math.cos( phi );
-                offset.y = 100;
+                offset.y = 100 * scale;
                offset.z = radius * Math.sin( phi ) * Math.cos( theta );
 
                // rotate offset back to "camera-up-vector-is-up" space
@@ -327,7 +327,6 @@
 
                thetaDelta = 0;
                phiDelta = 0;
-               scale = 1;
                pan.set( 0, 0, 0 );
 
                // update condition is:
diff -r -U 3 ff7.orig/jrra.zone.html ff7/jrra.zone.html
--- ff7.orig/jrra.zone.html     2016-03-29 12:22:51.000000000 -0700
+++ ff7/jrra.zone.html  2017-12-31 16:44:19.615000551 -0800
@@ -331,7 +331,7 @@
             controls = new MapControls( camera, renderer.domElement );
             controls.enableDamping = true;
             controls.dampingFactor = 0.25;
-            controls.enableZoom = false;
+            controls.noZoom = false;
             var shouldUpdate = false;
             controls.addEventListener("change", function() {
                 shouldUpdate = true;
@@ -585,4 +585,4 @@
         </script>
         
     </body>
-</html>
\ No newline at end of file
+</html>