Index: jalv/src/jalv_gtk.c
===================================================================
--- jalv/src/jalv_gtk.c	(Revision 5707)
+++ jalv/src/jalv_gtk.c	(Arbeitskopie)
@@ -191,6 +191,18 @@
 
 		active_preset_item = GTK_CHECK_MENU_ITEM(widget);
 		gtk_check_menu_item_set_active(active_preset_item, TRUE);
+	
+		// Add preset name to window tittle
+		LilvNode* name = lilv_plugin_get_name(record->jalv->plugin);
+		const char * tittle = lilv_node_as_string(name);
+		lilv_node_free(name);
+		const char * preset = gtk_menu_item_get_label(GTK_MENU_ITEM(widget));
+		char* new_tittle = calloc(strlen(tittle)+strlen(" - ")+strlen(preset)+1, sizeof(char)); 
+		strcpy(new_tittle,tittle);
+		strcat(new_tittle," - ");
+		strcat(new_tittle,preset);
+		gtk_window_set_title(GTK_WINDOW(record->jalv->window), new_tittle);
+		free(new_tittle);
 	}
 }
 
@@ -398,6 +410,17 @@
 		// Rebuild preset menu
 		rebuild_preset_menu(jalv, GTK_CONTAINER(gtk_widget_get_parent(widget)));
 
+		// Add preset name to window tittle
+		LilvNode* name = lilv_plugin_get_name(jalv->plugin);
+		const char * tittle = lilv_node_as_string(name);
+		lilv_node_free(name);
+		char* new_tittle = calloc(strlen(tittle)+strlen(" - ")+strlen(basename)+1, sizeof(char)); 
+		strcpy(new_tittle,tittle);
+		strcat(new_tittle," - ");
+		strcat(new_tittle,basename);
+		gtk_window_set_title(GTK_WINDOW(jalv->window), new_tittle);
+		free(new_tittle);
+
 		g_free(dir);
 		g_free(file);
 		g_free(bundle);
@@ -438,6 +461,12 @@
 		rebuild_preset_menu(jalv, GTK_CONTAINER(gtk_widget_get_parent(widget)));
 	}
 
+	// Set plugin name as window tittle
+	LilvNode* name = lilv_plugin_get_name(jalv->plugin);
+	const char * tittle = lilv_node_as_string(name);
+	lilv_node_free(name);
+	gtk_window_set_title(GTK_WINDOW(jalv->window), tittle);
+
 	g_free(msg);
 	gtk_widget_destroy(text);
 	gtk_widget_destroy(dialog);