Index: binaries/data/mods/public/gui/replaymenu/replay_actions.js
===================================================================
--- binaries/data/mods/public/gui/replaymenu/replay_actions.js	(revision 21022)
+++ binaries/data/mods/public/gui/replaymenu/replay_actions.js	(working copy)
@@ -125,16 +125,32 @@ function showReplaySummary()
 		"sim": simData,
 		"gui": {
 			"dialog": false,
 			"isReplay": true,
 			"replayDirectory": g_ReplaysFiltered[selected].directory,
-			"replaySelectionData": createReplaySelectionData(g_ReplaysFiltered[selected].directory)
+			"replaySelectionData": createReplaySelectionData(g_ReplaysFiltered[selected].directory),
+			"next": nextSummary(selected, 1),
+			"previous": nextSummary(selected, -1)
 		},
 		"selectedData": g_SummarySelectedData
 	});
 }
 
+function nextSummary(index, direction)
+{
+	let nextIdx = index;
+	do
+	{
+		nextIdx += direction;
+		if (nextIdx >= g_ReplaysFiltered.length || nextIdx < 0)
+			return -1;
+		if (Engine.HasReplayMetadata(g_ReplaysFiltered[nextIdx].directory))
+			break;
+	} while (true);
+	return nextIdx;
+}
+
 function reloadCache()
 {
 	let selected = Engine.GetGUIObjectByName("replaySelection").selected;
 	loadReplays(selected > -1 ? createReplaySelectionData(g_ReplaysFiltered[selected].directory) : "", true);
 }
Index: binaries/data/mods/public/gui/replaymenu/replay_menu.js
===================================================================
--- binaries/data/mods/public/gui/replaymenu/replay_menu.js	(revision 21022)
+++ binaries/data/mods/public/gui/replaymenu/replay_menu.js	(working copy)
@@ -75,10 +75,16 @@ function init(data)
 	initHotkeyTooltips();
 	displayReplayList();
 
 	if (data && data.summarySelectedData)
 		g_SummarySelectedData = data.summarySelectedData;
+
+	if (data && data.showNextSummary !== undefined)
+	{
+		Engine.GetGUIObjectByName("replaySelection").selected = data.showNextSummary;
+		showReplaySummary();
+	}
 }
 
 /**
  * Store the list of replays loaded in C++ in g_Replays.
  * Check timestamp and compatibility and extract g_Playernames, g_MapNames, g_VictoryConditions.
Index: binaries/data/mods/public/gui/summary/summary.js
===================================================================
--- binaries/data/mods/public/gui/summary/summary.js	(revision 21022)
+++ binaries/data/mods/public/gui/summary/summary.js	(working copy)
@@ -164,10 +164,29 @@ function init(data)
 	{
 		let tab = Engine.GetGUIObjectByName(button);
 		tab.onMouseWheelUp = () => selectNextTab(1);
 		tab.onMouseWheelDown = () => selectNextTab(-1);
 	}
+
+	if (g_GameData.gui.previous !== undefined)
+	{
+		Engine.GetGUIObjectByName("prevReplayButton").hidden = false;
+		Engine.GetGUIObjectByName("nextReplayButton").hidden = false;
+		if (g_GameData.gui.previous == -1)
+			Engine.GetGUIObjectByName("prevReplayButton").enabled = false;
+		if (g_GameData.gui.next == -1)
+			Engine.GetGUIObjectByName("nextReplayButton").enabled = false;
+	}
+}
+
+function nextReplaySummary(nextSummary)
+{
+	if (g_GameData.gui.isReplay)
+		Engine.SwitchGuiPage("page_replaymenu.xml", {
+			"replaySelectionData": g_GameData.gui.replaySelectionData,
+			"showNextSummary": nextSummary
+		});
 }
 
 /**
  * Sets the style and title of the page.
  */
Index: binaries/data/mods/public/gui/summary/summary.xml
===================================================================
--- binaries/data/mods/public/gui/summary/summary.xml	(revision 21022)
+++ binaries/data/mods/public/gui/summary/summary.xml	(working copy)
@@ -213,10 +213,20 @@
 			</repeat>
 
 			<object name="chartLegend" type="text" style="ModernLabelText" size="15 100%-35 100%-15 100%-5"/>
 		</object>
 
+		<object type="button" name="prevReplayButton" style="ModernButtonRed" size="20 100%-48 160 100%-20" hidden="true">
+			<translatableAttribute id="caption">Previous</translatableAttribute>
+			<action on="Press">nextReplaySummary(g_GameData.gui.previous)</action>
+		</object>
+
+		<object type="button" name="nextReplayButton" style="ModernButtonRed" size="170 100%-48 170+140 100%-20" hidden="true">
+			<translatableAttribute id="caption">Next</translatableAttribute>
+			<action on="Press">nextReplaySummary(g_GameData.gui.next)</action>
+		</object>
+
 		<object type="button" name="replayButton" style="ModernButtonRed" size="100%-310 100%-48 100%-170 100%-20">
 			<translatableAttribute id="caption">Replay</translatableAttribute>
 			<action on="Press">confirmStartReplay();</action>
 		</object>