diff --git a/src/server/game/Accounts/AccountMgr.cpp b/src/server/game/Accounts/AccountMgr.cpp
index 1737ab3..7be7a0d 100644
--- a/src/server/game/Accounts/AccountMgr.cpp
+++ b/src/server/game/Accounts/AccountMgr.cpp
@@ -83,7 +83,7 @@ AccountOpResult AccountMgr::DeleteAccount(uint32 accountId)
 
     trans->PAppend("DELETE FROM account WHERE id = '%d'", accountId);
     trans->PAppend("DELETE FROM account_access WHERE id = '%d'", accountId);
-    trans->PAppend("DELETE FROM realmcharacters WHERE accountId = '%d'", accountId);
+    trans->PAppend("DELETE FROM realmcharacters WHERE acctid = '%d'", accountId);
 
     LoginDatabase.CommitTransaction(trans);