Index: CCmpVisualActor.cpp
===================================================================
--- CCmpVisualActor.cpp	(revision 21013)
+++ CCmpVisualActor.cpp	(working copy)
@@ -20,6 +20,9 @@
 #include "simulation2/system/Component.h"
 #include "ICmpVisual.h"
 
+#include "scriptinterface/ScriptTypes.h"
+#include "scriptinterface/ScriptInterface.h"
+
 #include "simulation2/MessageTypes.h"
 
 #include "ICmpFootprint.h"
@@ -441,9 +444,11 @@
 		m_AnimSyncOffsetTime = fixed::Zero();
 
 		SetVariant("animation", m_AnimName);
+		CStr code = "let cmp = Engine.QueryInterface(" + CStr::FromUInt(m_Unit->GetID()) + ", IID_Sound); " + "if (cmp) cmp.GetSoundGroup('" + m_AnimName + "');";
+		GetSimContext().GetScriptInterface().Eval(code.c_str(), soundgroup);
 
 		if (m_Unit && m_Unit->GetAnimation())
-			m_Unit->GetAnimation()->SetAnimationState(m_AnimName, m_AnimOnce, m_AnimSpeed.ToFloat(), m_AnimDesync.ToFloat(), m_SoundGroup.c_str());
+			m_Unit->GetAnimation()->SetAnimationState(m_AnimName, m_AnimOnce, m_AnimSpeed.ToFloat(), m_AnimDesync.ToFloat(), soundgroup);
 	}
 
 	virtual void ReplaceMoveAnimation(const std::string& name, const std::string& replace)