Index: source/tools/i18n/updateTemplates.py
===================================================================
--- source/tools/i18n/updateTemplates.py	(revision 16919)
+++ source/tools/i18n/updateTemplates.py	(working copy)
@@ -70,10 +70,13 @@
 
     for templateSettings in settings:
         if "skip" in templateSettings and templateSettings["skip"] == "yes":
             continue
 
+        if templateSettings["output"] != "public-simulation-other.pot":
+            continue
+        
         inputRootPath = rootPath
         if "inputRoot" in templateSettings:
             inputRootPath = os.path.join(rootPath, templateSettings["inputRoot"])
 
         template = Catalog(os.path.join(rootPath, templateSettings["output"]), create=True, truncate=True)