diff --git a/rules/base.extras.xml.in b/rules/base.extras.xml.in
index e3af738..9fa8500 100644
--- a/rules/base.extras.xml.in
+++ b/rules/base.extras.xml.in
@@ -870,6 +870,16 @@
             </languageList>
           </configItem>
         </variant>
+        <variant>
+          <configItem>
+            <name>scn</name>
+            <_description>Sicilian</_description>
+            <languageList>
+              <iso639Id>it</iso639Id>
+              <iso639Id>scn</iso639Id>
+            </languageList>
+          </configItem>
+        </variant>
       </variantList>
     </layout>
     <layout>
diff --git a/rules/base.xml.in b/rules/base.xml.in
index 5e67937..6e86c74 100644
--- a/rules/base.xml.in
+++ b/rules/base.xml.in
@@ -3777,6 +3777,25 @@
             <_description>Italian (IBM 142)</_description>
           </configItem>
         </variant>
+        <variant>
+          <configItem>
+            <name>intl</name>
+            <_description>Italian (intl., with dead keys)</_description>
+            <languageList>
+              <iso639Id>de</iso639Id>
+              <iso639Id>fr</iso639Id>
+              <iso639Id>it</iso639Id>
+              <iso639Id>sl</iso639Id>
+              <iso639Id>sc</iso639Id>
+              <iso639Id>nap</iso639Id>
+              <iso639Id>scn</iso639Id>
+              <iso639Id>fur</iso639Id>
+              <iso639Id>lmo</iso639Id>
+              <iso639Id>aae</iso639Id>
+              <iso639Id>pms</iso639Id>
+            </languageList>
+          </configItem>
+        </variant>
       </variantList>
     </layout>
     <layout>
diff --git a/symbols/it b/symbols/it
index 11ee0aa..57edc39 100644
--- a/symbols/it
+++ b/symbols/it
@@ -367,3 +367,54 @@ partial alphanumeric_keys
     key <AB03> { [ c, C, cacute, Cacute ] };
 
 };
+
+// made by David Paleino - Cadèmia Siciliana <l10n@cademiasiciliana.org>
+partial alphanumeric_keys
+    xkb_symbols "intl" {
+    include "it(basic)"
+    name[Group1] = "Italian (intl., with dead keys)";
+
+    key <AE12> { [ iacute, asciicircum,  asciitilde, dead_circumflex ] };
+
+    key <AD03> { [      e,           E,    EuroSign,           U0259 ] };
+
+    key <AC11> { [ agrave,      degree,  numbersign,           U1D58 ] };
+    key <AC06> { [      h,           H,       U1E25,           U1E24 ] };
+    key <AC04> { [      f,           F,       U0111,           U0110 ] };
+    key <AC03> { [      d,           D,       U1E0D,           U1E0C ] };
+
+    key <AB02> { [      x,           X,       U03C7,           U03A7 ] };
+    key <AB01> { [      z,           Z                               ] };
+
+    key <BKSL> { [ uacute,     section,  dead_grave,      dead_breve ] };
+};
+
+partial alphanumeric_keys
+    xkb_symbols "scn" {
+    include "it(basic)"
+    name[Group1] = "Sicilian";
+
+    key <AE12> { [ iacute, asciicircum,  asciitilde, dead_circumflex ] };
+
+    key <AD09> { [      o,           O, ocircumflex,     Ocircumflex ] };
+    key <AD08> { [      i,           I, icircumflex,     Icircumflex ] };
+    key <AD07> { [      u,           U, ucircumflex,     Ucircumflex ] };
+    key <AD04> { [      r,           R,  ediaeresis,      Ediaeresis ] };
+    key <AD03> { [      e,           E,    EuroSign,     ecircumflex ] };
+
+    key <AC11> { [ agrave,      degree,  numbersign,           U1D58 ] };
+    key <AC04> { [      f,           F,       U0111,           U0110 ] };
+    key <AC03> { [      d,           D,       U1E0D,           U1E0C ] };
+    key <AC02> { [      s,           S,      scaron,          Scaron ] };
+    key <AC01> { [      a,           A, acircumflex,     Acircumflex ] };
+
+    // In our Windows layout, there is no ©, but Ç. Since this character
+    // is available via CapsLock, I decided to keep © in place, as the
+    // Italian layout.
+    key <AB03> { [      c,           C,    ccedilla,       copyright ] };
+    key <AB02> { [      x,           X,       U03C7,           U03A7 ] };
+
+    // Replaced dead_grave by ugrave, since it's common to write mixed
+    // Sicilian - Italian (where we replaced ugrave with uacute)
+    key <BKSL> { [ uacute,     section,      ugrave,      dead_breve ] };
+};
diff --git a/symbols/us b/symbols/us
index f488a8e..d1e3533 100644
--- a/symbols/us
+++ b/symbols/us
@@ -1642,6 +1642,22 @@ xkb_symbols "cz_sk_de" {
     include "level3(ralt_switch)"
 };
 
+partial alphanumeric_keys
+xkb_symbols "scn" {
+
+    include "us(intl)"
+    name[Group1]="Sicilian (US keyboard)";
+
+    key <AD03> { [      e,       E, U0259,       U018F ] };
+    key <AC03> { [      d,       D, U1E0D,       U1E0C ] };
+    key <AC04> { [      f,       F, U0111,       U0110 ] };
+    key <AC06> { [      h,       H, U1E25,       U1E24 ] };
+    key <AB02> { [      x,       X, U03C7,       U03A7 ] };
+    key <AB09> { [ period, greater, U1D58,  dead_caron ] };
+
+    include "level3(ralt_switch)"
+};
+
 // XCompose is out! Unicode combining is in!  For those of us who live
 // on the edge: A keymap using Unicode combining characters instead of
 // deadkeys.  This variation does not deviate from the lame MS-style