diff --git a/1.js b/1.js
index 59c9300e63cb89321962fb58cd0d6d33ba050138..16865795c124ddee478a151a13ff9a813fd05a5f 100644
--- a/1.js
+++ b/1.js
@@ -1,20 +1,28 @@
+/*
+* File: jQuery loader
+* Author: Ben Frain
+* Project: MacUser
+* Detail: File changes with Subversion (SVN)
+*/
+
 $().ready(function() {
   $(".linksTrigger").hover(function(){ // .linksTrigger class is used on a <a href="#"> tag
   $(this).toggleClass("switch"); // this bit adds a class of 'switch' to the trigger element so it can be styled differently than the default state
   $(".extraLink").toggle("fast"); // this bit finds any elements with a class of .extraLink and toggle's them
-  $(this).text($(this).text() == 'Click for more Site Links' ? 'Click for less Site Links' : 'Click for more Site Links'); // this is the text that will toggle when clicked
-  return false; //Prevent the browser jump to the link anchor
+  $(this).text($(this).text() == 'More Site Links' ? 'Less Site Links' : 'More Site Links'); // this is the text that will toggle when clicked
+  
   });
 });
 
 jQuery(document).ready(function(){
 // Add Cufon
-Cufon.replace('.MUHeaderCatTitle, h2, .latestnews_title');
+Cufon.replace('.MUHeaderCatTitle, h2, .latestnews_title, .p');
 });
 //Start slideshow cycle plugin
 jQuery('#slideshow').cycle({
         prev:   '#prev',
         next:   '#next',
+        fade: 'fast'
         timeout: 8000
     });