diff --git a/acls.c b/acls.c
index 0f98df2..efb9530 100644
--- a/acls.c
+++ b/acls.c
@@ -178,7 +178,7 @@ struct acluser **up;
 #endif
   (*up)->u_Esc = DefaultEsc;
   (*up)->u_MetaEsc = DefaultMetaEsc;
-  strncpy((*up)->u_name, name, 20);
+  strncpy((*up)->u_name, name, MAXLOGINLEN);
   (*up)->u_password = NULL;
   if (pass)
     (*up)->u_password = SaveStr(pass);
diff --git a/acls.h b/acls.h
index c41b714..1241e31 100644
--- a/acls.h
+++ b/acls.h
@@ -73,7 +73,7 @@ struct plop
 typedef struct acluser
 {
   struct acluser *u_next;		/* continue the main user list */
-  char u_name[20+1];		/* login name how he showed up */
+  char u_name[MAXLOGINLEN + 1];		/* login name how he showed up */
   char *u_password;		/* his password (may be NullStr). */
   int  u_checkpassword;		/* nonzero if this u_password is valid */
   int  u_detachwin;		/* the window where he last detached */
diff --git a/comm.c b/comm.c
index 0934847..96a8315 100644
--- a/comm.c
+++ b/comm.c
@@ -31,6 +31,7 @@
  */
 
 #include "config.h"
+#include "os.h"
 #include "acls.h"
 #include "comm.h"
 
diff --git a/display.h b/display.h
index ef99954..ec92dba 100644
--- a/display.h
+++ b/display.h
@@ -85,7 +85,7 @@ struct display
   struct win *d_other;		/* pointer to other window */
   int   d_nonblock;		/* -1 don't block if obufmax reached */
 				/* >0: block after nonblock secs */
-  char  d_termname[20 + 1];	/* $TERM */
+  char  d_termname[MAXTERMLEN + 1];	/* $TERM */
   char	*d_tentry;		/* buffer for tgetstr */
   char	d_tcinited;		/* termcap inited flag */
   int	d_width, d_height;	/* width/height of the screen */
diff --git a/os.h b/os.h
index 7cb1352..af4c1d0 100644
--- a/os.h
+++ b/os.h
@@ -531,3 +531,8 @@ typedef struct fd_set { int fds_bits[1]; } fd_set;
  */
 #define IOSIZE		4096
 
+/* Changing those you won't be able to attach to your old sessions
+ * when changing those values in official tree don't forget to bump
+ * MSG_VERSION */
+#define MAXTERMLEN  40
+#define MAXLOGINLEN 256
diff --git a/process.c b/process.c
index 7e225f0..6437a66 100644
--- a/process.c
+++ b/process.c
@@ -2442,9 +2442,9 @@ int key;
       s = NULL;
       if (ParseSaveStr(act, &s))
 	break;
-      if (strlen(s) >= 20)
+      if (strlen(s) >= MAXTERMLEN)
 	{
-	  Msg(0, "%s: term: argument too long ( < 20)", rc_name);
+	  Msg(0, "%s: term: argument too long ( < %d)", rc_name, MAXTERMLEN);
 	  free(s);
 	  break;
 	}
diff --git a/screen.c b/screen.c
index fab028d..2817c55 100644
--- a/screen.c
+++ b/screen.c
@@ -889,10 +889,10 @@ char **av;
 
   if (home == 0 || *home == '\0')
     home = ppp->pw_dir;
-  if (strlen(LoginName) > 20)
+  if (strlen(LoginName) > MAXLOGINLEN)
     Panic(0, "LoginName too long - sorry.");
 #ifdef MULTIUSER
-  if (multi && strlen(multi) > 20)
+  if (multi && strlen(multi) > MAXLOGINLEN)
     Panic(0, "Screen owner name too long - sorry.");
 #endif
   if (strlen(home) > MAXPATHLEN - 25)
diff --git a/screen.h b/screen.h
index 5e8e59a..963392b 100644
--- a/screen.h
+++ b/screen.h
@@ -190,31 +190,31 @@ struct msg
 	  int nargs;
 	  char line[MAXPATHLEN];
 	  char dir[MAXPATHLEN];
-	  char screenterm[20];	/* is screen really "screen" ? */
+	  char screenterm[MAXTERMLEN];	/* is screen really "screen" ? */
 	}
       create;
       struct
 	{
-	  char auser[20 + 1];	/* username */
+	  char auser[MAXLOGINLEN + 1];	/* username */
 	  int apid;		/* pid of frontend */
 	  int adaptflag;	/* adapt window size? */
 	  int lines, columns;	/* display size */
 	  char preselect[20];
 	  int esc;		/* his new escape character unless -1 */
 	  int meta_esc;		/* his new meta esc character unless -1 */
-	  char envterm[20 + 1];	/* terminal type */
+	  char envterm[MAXTERMLEN + 1];	/* terminal type */
 	  int encoding;		/* encoding of display */
 	}
       attach;
       struct 
 	{
-	  char duser[20 + 1];	/* username */
+	  char duser[MAXLOGINLEN + 1];	/* username */
 	  int dpid;		/* pid of frontend */
 	}
       detach;
       struct 
 	{
-	  char auser[20 + 1];	/* username */
+	  char auser[MAXLOGINLEN + 1];	/* username */
 	  int nargs;
 	  char cmd[MAXPATHLEN];	/* command */
 	  int apid;		/* pid of frontend */
diff --git a/socket.c b/socket.c
index e720f6e..c1f7505 100644
--- a/socket.c
+++ b/socket.c
@@ -1248,7 +1248,7 @@ static void PasswordProcessInput __P((char *, int));
 
 struct pwdata {
   int l;
-  char buf[20 + 1];
+  char buf[MAXLOGINLEN + 1];
   struct msg m;
 };