--- en07Y90f    2017-11-30 10:39:04.463068497 +0100
+++ en07Y90f.asm        2017-11-30 10:38:55.319108871 +0100
@@ -40,6 +40,7 @@
     syscall

     mov rdi, password
+    mov rdx, password_length

     call compare

@@ -50,15 +51,13 @@
     jmp _end

 compare:
-    cmp rdi, rsi
+    cmpsb
     jnz compare_invalid

-    cmp rdx, 0
+    dec rdx
+    ; cmp rdx, 0 ; ten cmp mozna sobie darowac, dec tez ustawia flage jak bedzie 0     jz compare_ret

-    dec rdx
-    inc rdi
-    inc rsi
     jmp compare

 compare_invalid: