diff --git a/packages/tools/installer/scripts/installer b/packages/tools/installer/scripts/installer
index 4858d4b..8fbdfc1 100755
--- a/packages/tools/installer/scripts/installer
+++ b/packages/tools/installer/scripts/installer
@@ -275,6 +275,8 @@ do_install_quick() {
           cp $TMPDIR/part1/syslinux.cfg $TMPDIR/part1/EFI/BOOT
           cp /usr/share/syslinux/bootx64.efi $TMPDIR/part1/EFI/BOOT
           cp /usr/share/syslinux/ldlinux.e64 $TMPDIR/part1/EFI/BOOT
+          cp /usr/share/syslinux/bootia32.efi $TMPDIR/part1/EFI/BOOT
+          cp /usr/share/syslinux/ldlinux.e32 $TMPDIR/part1/EFI/BOOT
         fi
         sync
 
diff --git a/packages/tools/syslinux/package.mk b/packages/tools/syslinux/package.mk
index 0538bed..10ce83d 100644
--- a/packages/tools/syslinux/package.mk
+++ b/packages/tools/syslinux/package.mk
@@ -77,6 +77,8 @@ makeinstall_host() {
     cp bios/mbr/gptmbr.bin $TOOLCHAIN/share/syslinux
     cp efi64/efi/syslinux.efi $TOOLCHAIN/share/syslinux/bootx64.efi
     cp efi64/com32/elflink/ldlinux/ldlinux.e64  $TOOLCHAIN/share/syslinux
+    cp efi32/efi/syslinux.efi $TOOLCHAIN/share/syslinux/bootia32.efi
+    cp efi32/com32/elflink/ldlinux/ldlinux.e32  $TOOLCHAIN/share/syslinux
 }
 
 makeinstall_target() {
@@ -92,4 +94,6 @@ makeinstall_target() {
     cp bios/mbr/gptmbr.bin $INSTALL/usr/share/syslinux
     cp efi64/efi/syslinux.efi $INSTALL/usr/share/syslinux/bootx64.efi
     cp efi64/com32/elflink/ldlinux/ldlinux.e64  $INSTALL/usr/share/syslinux
+    cp efi32/efi/syslinux.efi $INSTALL/usr/share/syslinux/bootia32.efi
+    cp efi32/com32/elflink/ldlinux/ldlinux.e32  $INSTALL/usr/share/syslinux
 }
diff --git a/scripts/mkimage b/scripts/mkimage
index 6ff5b5e..dfe6a12 100755
--- a/scripts/mkimage
+++ b/scripts/mkimage
@@ -84,7 +84,7 @@ trap cleanup SIGINT
   SYSTEM_PART_END=$(( $SYSTEM_PART_START + ($SYSTEM_SIZE * 1024 * 1024 / 512) - 1 ))
   parted -s "$DISK" -a min unit s mkpart primary fat32 $SYSTEM_PART_START $SYSTEM_PART_END
   if [ "$BOOTLOADER" = "syslinux" ]; then
-    parted -s "$DISK" set 1 legacy_boot on
+    parted -s "$DISK" set 1 boot on
   else
     parted -s "$DISK" set 1 boot on
   fi
@@ -163,6 +163,8 @@ EOF
     mmd EFI EFI/BOOT
     mcopy $TOOLCHAIN/share/syslinux/bootx64.efi ::/EFI/BOOT
     mcopy $TOOLCHAIN/share/syslinux/ldlinux.e64 ::/EFI/BOOT
+    mcopy $TOOLCHAIN/share/syslinux/bootia32.efi ::/EFI/BOOT
+    mcopy $TOOLCHAIN/share/syslinux/ldlinux.e32 ::/EFI/BOOT
     mcopy "$LE_TMP"/syslinux.cfg ::/EFI/BOOT
 elif [ "$BOOTLOADER" = "bcm2835-bootloader" ]; then
   # create bootloader configuration