diff --git a/src/pregmod/widgets/economyWidgets.tw b/src/pregmod/widgets/economyWidgets.tw
index 5e819c6..0e4a3ed 100644
--- a/src/pregmod/widgets/economyWidgets.tw
+++ b/src/pregmod/widgets/economyWidgets.tw
@@ -399,10 +399,10 @@
 			<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 			<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 			<<replace '#cashOnHand'>>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace `'#'+$args[0]+'AssetsDisp'`>>
-				¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+				<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 			<</replace>>
 			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 		<<else>>
@@ -419,10 +419,10 @@
 		<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 		<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print cashFormat($corpCash)>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Successful buy */
@@ -430,10 +430,10 @@
 		<<set State.variables[$args[0]+"Assets"] += (eval(parse('_Num' + $args[0]))) * 500>>
 		<<set $corpCash -= (eval(parse('_Num' + $args[0]))) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print cashFormat($corpCash)>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<print Math.floor(State.variables[$args[0]+"Assets"]/500)>> unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>><</replace>>
 /* Unsuccessful buy but have enough cash for more than 1 unit */
@@ -459,10 +459,10 @@
 			<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 			<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 			<<replace '#cashOnHand'>>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace `'#'+$args[0]+'AssetsDisp'`>>
-				¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+				<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 			<</replace>>
 			<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Error if decimal or not "all" or "max" string */
@@ -479,10 +479,10 @@
 		<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 		<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print cashFormat($corpCash)>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<if Math.ceil(State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+'controls'+$args[0]`>><<CorpDevBuySell $args[0] $args[1]>><</replace>><<script>>setReplaceTextboxMaxLength(State.temporary["textboxMLArg"], 10);<</script>>
 /* Successful sell */
@@ -490,10 +490,10 @@
 		<<set State.variables[$args[0]+"Assets"] -= eval(parse('_Num' + $args[0])) * 500>>
 		<<set $corpCash += eval(parse('_Num' + $args[0])) * 500 * State.variables[$args[0]+"AssetPrice"]>>
 		<<replace '#cashOnHand'>>
-			¤<<print $corpCash>>
+			<<print cashFormat($corpCash)>>
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsDisp'`>>
-			¤<<print State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"]>> (<<if (State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
+			<<print cashFormat(State.variables[$args[0]+"AssetPrice"] * State.variables[$args[0]+"Assets"])>> (<<if (State.variables[$args[0]+"Assets"]/500) <1>> <1 <<else>><<print Math.ceil(State.variables[$args[0]+"Assets"]/500)>> <</if>>unit<<if State.variables[$args[0]+"Assets"]/500 > 1>>s<</if>>)
 		<</replace>>
 		<<replace `'#'+$args[0]+'AssetsPurchasing'`>><</replace>>
 /* Unsuccessful sell but have enough assets for more than 1 unit */
diff --git a/src/uncategorized/corporationDevelopments.tw b/src/uncategorized/corporationDevelopments.tw
index 33291b0..cc830b7 100644
--- a/src/uncategorized/corporationDevelopments.tw
+++ b/src/uncategorized/corporationDevelopments.tw
@@ -216,7 +216,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="genAssets">¤<<print $generalAssetPrice*$generalAssets>></span> in general assets.
+	<span id="genAssets"><<print cashFormat($generalAssetPrice*$generalAssets)>></span> in general assets.
 	General asset prices are
 	<<if $generalAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $generalAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $generalAssetPrice == 10>>average<<elseif $generalAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -224,10 +224,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $generalAssets += 500>>
 			<<set $corpCash -= 500*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -242,10 +242,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $generalAssets += (Math.trunc($corpCash/($generalAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($generalAssetPrice*500)))*($generalAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 		<<else>>
 		<<replace "#purchasing">>
@@ -258,10 +258,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $generalAssets -= 500>>
 			<<set $corpCash += 500*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -278,10 +278,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $generalAssets = Math.trunc($generalAssets/2)>>
 			<<set $corpCash += $generalAssets*$generalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#genAssets">>
-				¤<<print $generalAssetPrice*$generalAssets>>
+				<<print cashFormat($generalAssetPrice*$generalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -295,11 +295,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="generalAssetsDisp"><<print cashFormat($generalAssetPrice*$generalAssets)>> (<<print commaNum($generalAssets/500)>> unit<<if $generalAssets/500 > 1>>s<</if>>)</span> in general assets.
 	Prices are 
 	<<if $generalAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $generalAssetPrice*500>>ea@@.
+		@@.yellowgreen;<<print cashFormat($generalAssetPrice*500)>>ea@@.
 	<<elseif $generalAssetPrice < 10>>
-		@@.gold;¤<<print $generalAssetPrice*500>>ea@@.
+		@@.gold;<<print cashFormat($generalAssetPrice*500)>>ea@@.
 	<<else>>
-		¤<<print $generalAssetPrice*500>>ea.
+		<<print cashFormat($generalAssetPrice*500)>>ea.
 	<</if>>
 	<span id="controlsgeneral"><<CorpDevBuySell "general" "Numgeneral">></span><span id='generalAssetsPurchasing'></span>
 <</if>>
@@ -308,7 +308,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <<set $slaveAssetPrice = Math.trunc(10*$slaveCostFactor)>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="slavAssets">¤<<print $slaveAssetPrice*$slaveAssets>></span> in slaves.
+	<span id="slavAssets"><<print cashFormat($slaveAssetPrice*$slaveAssets)>></span> in slaves.
 	Slave prices are
 	<<if $slaveAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $slaveAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $slaveAssetPrice == 10>>average<<elseif $slaveAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -317,10 +317,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $corpPeopleEnslaved += Math.trunc((500*$slaveAssetPrice)/1000)>>
 			<<set $corpCash -= 500*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -336,10 +336,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $corpPeopleEnslaved += Math.trunc(($corpCash/($slaveAssetPrice*2))/1000)>>
 			<<set $corpCash -= (Math.trunc($corpCash/($slaveAssetPrice*500)))*($slaveAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -352,10 +352,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $slaveAssets -= 500>>
 			<<set $corpCash += 500*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -370,10 +370,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $slaveAssets = Math.trunc($slaveAssets/2)>>
 			<<set $corpCash += $slaveAssets*$slaveAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#slavAssets">>
-				¤<<print $slaveAssetPrice*$slaveAssets>>
+				<<print cashFormat($slaveAssetPrice*$slaveAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -387,11 +387,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="slaveAssetsDisp"><<print cashFormat($slaveAssetPrice*$slaveAssets)>> (<<print commaNum(Math.floor($slaveAssets/500))>> unit<<if Math.floor($slaveAssets/500) > 1>>s<</if>>)</span> in slaves.
 	Prices are 
 	<<if $slaveAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $slaveAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($slaveAssetPrice*500)>>ea@@
 	<<elseif $slaveAssetPrice < 10>>
-		@@.gold;¤<<print $slaveAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($slaveAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $slaveAssetPrice*500>>ea
+		<<print cashFormat($slaveAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsslave"><<CorpDevBuySell "slave" "Numslave">></span><span id='slaveAssetsPurchasing'></span>
 <</if>>
@@ -422,7 +422,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="trapAssets">¤<<print $entrapmentAssetPrice*$entrapmentAssets>></span> in legal enslavement assets.
+	<span id="trapAssets"><<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>></span> in legal enslavement assets.
 	Legal enslavement asset prices are
 	<<if $entrapmentAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $entrapmentAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $entrapmentAssetPrice == 10>>average<<elseif $entrapmentAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -430,10 +430,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $entrapmentAssets += 500>>
 			<<set $corpCash -= 500*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -448,10 +448,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $entrapmentAssets += (Math.trunc($corpCash/($entrapmentAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($entrapmentAssetPrice*500)))*($entrapmentAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -464,10 +464,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $entrapmentAssets -= 500>>
 			<<set $corpCash += 500*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -482,10 +482,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $entrapmentAssets = Math.trunc($entrapmentAssets/2)>>
 			<<set $corpCash += $entrapmentAssets*$entrapmentAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trapAssets">>
-				¤<<print $entrapmentAssetPrice*$entrapmentAssets>>
+				<<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -499,11 +499,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="entrapmentAssetsDisp"><<print cashFormat($entrapmentAssetPrice*$entrapmentAssets)>> (<<print commaNum(Math.floor($entrapmentAssets/500))>> unit<<if Math.floor($entrapmentAssets/500) > 1>>s<</if>>)</span> in legal enslavement assets.
 	Prices are 
 	<<if $entrapmentAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $entrapmentAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($entrapmentAssetPrice*500)>>ea@@
 	<<elseif $entrapmentAssetPrice < 10>>
-		@@.gold;¤<<print $entrapmentAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($entrapmentAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $entrapmentAssetPrice*500>>ea
+		<<print cashFormat($entrapmentAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsentrapment"><<CorpDevBuySell "entrapment" "Numentrapment">></span><span id='entrapmentAssetsPurchasing'></span>
 <</if>>
@@ -534,7 +534,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="capAssets">¤<<print $captureAssetPrice*$captureAssets>></span> in conflict zone enslavement assets.
+	<span id="capAssets"><<print cashFormat($captureAssetPrice*$captureAssets)>></span> in conflict zone enslavement assets.
 	Extralegal enslavement asset prices are
 	<<if $captureAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $captureAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $captureAssetPrice == 10>>average<<elseif $captureAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -542,10 +542,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $captureAssets += 500>>
 			<<set $corpCash -= 500*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -560,10 +560,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $captureAssets += (Math.trunc($corpCash/($captureAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($captureAssetPrice*500)))*($captureAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -576,10 +576,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $captureAssets -= 500>>
 			<<set $corpCash += 500*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -594,10 +594,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $captureAssets = Math.trunc($captureAssets/2)>>
 			<<set $corpCash += $captureAssets*$captureAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#capAssets">>
-				¤<<print $captureAssetPrice*$captureAssets>>
+				<<print cashFormat($captureAssetPrice*$captureAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -611,11 +611,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="captureAssetsDisp"><<print cashFormat($captureAssetPrice*$captureAssets)>> (<<print commaNum(Math.floor($captureAssets/500))>> unit<<if Math.floor($captureAssets/500) > 1>>s<</if>>)</span> in conflict zone enslavement assets.
 	Prices are 
 	<<if $captureAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $captureAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($captureAssetPrice*500)>>ea@@
 	<<elseif $captureAssetPrice < 10>>
-		@@.gold;¤<<print $captureAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($captureAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $captureAssetPrice*500>>ea
+		<<print cashFormat($captureAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlscapture"><<CorpDevBuySell "capture" "Numcapture">></span><span id='captureAssetsPurchasing'></span>
 <</if>>
@@ -646,7 +646,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="trainAssets">¤<<print $trainingAssetPrice*$trainingAssets>></span> in slave training assets.
+	<span id="trainAssets"><<print cashFormat($trainingAssetPrice*$trainingAssets)>></span> in slave training assets.
 	Slave training asset prices are
 	<<if $trainingAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $trainingAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $trainingAssetPrice == 10>>average<<elseif $trainingAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -654,10 +654,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $trainingAssets += 500>>
 			<<set $corpCash -= 500*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -672,10 +672,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $trainingAssets += (Math.trunc($corpCash/($trainingAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($trainingAssetPrice*500)))*($trainingAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -688,10 +688,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $trainingAssets -= 500>>
 			<<set $corpCash += 500*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -706,10 +706,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $trainingAssets = Math.trunc($trainingAssets/2)>>
 			<<set $corpCash += $trainingAssets*$trainingAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#trainAssets">>
-				¤<<print $trainingAssetPrice*$trainingAssets>>
+				<<print cashFormat($trainingAssetPrice*$trainingAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -723,11 +723,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="trainingAssetsDisp"><<print cashFormat($trainingAssetPrice*$trainingAssets)>> (<<print commaNum(Math.floor($trainingAssets/500))>> unit<<if Math.floor($trainingAssets/500) > 1>>s<</if>>)</span> of slave training assets.
 	Prices are 
 	<<if $trainingAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $trainingAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($trainingAssetPrice*500)>>ea@@
 	<<elseif $trainingAssetPrice < 10>>
-		@@.gold;¤<<print $trainingAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($trainingAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $trainingAssetPrice*500>>ea
+		<<print cashFormat($trainingAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlstraining"><<CorpDevBuySell "training" "Numtraining">></span><span id='trainingAssetsPurchasing'></span>
 <</if>>
@@ -758,7 +758,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="surgAssets">¤<<print $surgicalAssetPrice*$surgicalAssets>></span> in surgical assets.
+	<span id="surgAssets"><<print cashFormat($surgicalAssetPrice*$surgicalAssets)>></span> in surgical assets.
 	Surgical asset prices are
 	<<if $surgicalAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $surgicalAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $surgicalAssetPrice == 10>>average<<elseif $surgicalAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -766,10 +766,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $surgicalAssets += 500>>
 			<<set $corpCash -= 500*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -784,10 +784,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $surgicalAssets += (Math.trunc($corpCash/($surgicalAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($surgicalAssetPrice*500)))*($surgicalAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -800,10 +800,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $surgicalAssets -= 500>>
 			<<set $corpCash += 500*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -818,10 +818,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $surgicalAssets = Math.trunc($surgicalAssets/2)>>
 			<<set $corpCash += $surgicalAssets*$surgicalAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#surgAssets">>
-				¤<<print $surgicalAssetPrice*$surgicalAssets>>
+				<<print cashFormat($surgicalAssetPrice*$surgicalAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -835,11 +835,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="surgicalAssetsDisp"><<print cashFormat($surgicalAssetPrice*$surgicalAssets)>> (<<print commaNum(Math.floor($surgicalAssets/500))>> unit<<if Math.floor($surgicalAssets/500) > 1>>s<</if>>)</span> in surgical assets.
 	Prices are 
 	<<if $surgicalAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $surgicalAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($surgicalAssetPrice*500)>>ea@@
 	<<elseif $surgicalAssetPrice < 10>>
-		@@.gold;¤<<print $surgicalAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($surgicalAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $surgicalAssetPrice*500>>ea
+		<<print cashFormat($surgicalAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlssurgical"><<CorpDevBuySell "surgical" "Numsurgical">></span><span id='surgicalAssetsPurchasing'></span>
 <</if>>
@@ -870,7 +870,7 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 <</if>>
 &nbsp;&nbsp;&nbsp;&nbsp;
 <<if $detailedCorpControls != 1>>
-	<span id="drAssets">¤<<print $drugAssetPrice*$drugAssets>></span> in pharmaceutical assets.
+	<span id="drAssets"><<print cashFormat($drugAssetPrice*$drugAssets)>></span> in pharmaceutical assets.
 	Drug asset prices are
 	<<if $drugAssetPrice > 12>>@@.yellowgreen;very high@@<<elseif $drugAssetPrice > 10>>@@.yellowgreen;high@@<<elseif $drugAssetPrice == 10>>average<<elseif $drugAssetPrice >= 8>>@@.gold;low@@<<else>>@@.gold;very low@@<</if>>.
 	| <<link "Buy">>
@@ -878,10 +878,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $drugAssets += 500>>
 			<<set $corpCash -= 500*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -896,10 +896,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $drugAssets += (Math.trunc($corpCash/($drugAssetPrice*500)))*500>>
 			<<set $corpCash -= (Math.trunc($corpCash/($drugAssetPrice*500)))*($drugAssetPrice*500)>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 		<<else>>
 			<<replace "#purchasing">>
@@ -912,10 +912,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $drugAssets -= 500>>
 			<<set $corpCash += 500*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -930,10 +930,10 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 			<<set $drugAssets = Math.trunc($drugAssets/2)>>
 			<<set $corpCash += $drugAssets*$drugAssetPrice>>
 			<<replace "#cashOnHand">>
-				¤<<print $corpCash>>
+				<<print cashFormat($corpCash)>>
 			<</replace>>
 			<<replace "#drAssets">>
-				¤<<print $drugAssetPrice*$drugAssets>>
+				<<print cashFormat($drugAssetPrice*$drugAssets)>>
 			<</replace>>
 			<<replace "#update">>
 			<</replace>>
@@ -947,11 +947,11 @@ You hold <<print commaNum($personalShares)>> shares personally while <<print com
 	<span id="drugAssetsDisp"><<print cashFormat($drugAssetPrice*$drugAssets)>> (<<print commaNum(Math.floor($drugAssets/500))>> unit<<if Math.floor($drugAssets/500) > 1>>s<</if>>)</span> in pharmaceutical assets.
 	Prices are 
 	<<if $drugAssetPrice > 10>>
-		@@.yellowgreen;¤<<print $drugAssetPrice*500>>ea@@
+		@@.yellowgreen;<<print cashFormat($drugAssetPrice*500)>>ea@@
 	<<elseif $drugAssetPrice < 10>>
-		@@.gold;¤<<print $drugAssetPrice*500>>ea@@
+		@@.gold;<<print cashFormat($drugAssetPrice*500)>>ea@@
 	<<else>>
-		¤<<print $drugAssetPrice*500>>ea
+		<<print cashFormat($drugAssetPrice*500)>>ea
 	<</if>>
 	<span id="controlsdrug"><<CorpDevBuySell "drug" "Numdrug">></span><span id='drugAssetsPurchasing'></span>
 <</if>>