diff --git a/chrome/browser/ui/virtual_keyboard/virtual_keyboard_manager.cc b/chrome/browser/ui/virtual_keyboard/virtual_keyboard_manager.cc
index e8a35a9..dd80721 100644
--- a/chrome/browser/ui/virtual_keyboard/virtual_keyboard_manager.cc
+++ b/chrome/browser/ui/virtual_keyboard/virtual_keyboard_manager.cc
@@ -26,6 +26,7 @@
 #include "ui/gfx/interpolated_transform.h"
 #include "ui/gfx/screen.h"
 #include "views/ime/text_input_type_tracker.h"
+#include "views/widget/native_widget_views.h"
 #include "views/widget/widget.h"
 
 #if defined(OS_CHROMEOS)
@@ -215,9 +216,9 @@ void KeyboardWidget::ShowKeyboardForWidget(views::Widget* widget) {
   transform_.reset(new ui::InterpolatedTranslation(
       gfx::Point(0, keyboard_height_), gfx::Point()));

-  GetRootView()->SetTransform(
-      transform_->Interpolate(animation_->GetCurrentValue()));
-  animation_->Show();
+//  static_cast<views::NativeWidgetViews*>(native_widget())->GetView()->SetTransform(
+//      transform_->Interpolate(animation_->GetCurrentValue()));
+//  animation_->Show();
   
   Show();

@@ -233,8 +234,9 @@ void KeyboardWidget::ResetBounds() {
 }
 
 void KeyboardWidget::Hide() {
-  animation_->Hide();
+//  animation_->Hide();

+  Widget::Hide();
   bool visible = false;
   NotificationService::current()->Notify(
       chrome::NOTIFICATION_KEYBOARD_VISIBILITY_CHANGED,
@@ -261,10 +263,11 @@ bool KeyboardWidget::OnKeyEvent(const views::KeyEvent& event) {
 }
 
 void KeyboardWidget::AnimationProgressed(const ui::Animation* animation) {
+  NOTREACHED();
   float t = static_cast<float>(animation_->GetCurrentValue());
-  if (GetRootView()->layer())
-    GetRootView()->layer()->SetOpacity(t * t);
-  GetRootView()->SetTransform(transform_->Interpolate(t));
+  if (static_cast<views::NativeWidgetViews*>(native_widget())->GetView()->layer())
+    static_cast<views::NativeWidgetViews*>(native_widget())->GetView()->layer()->SetOpacity(t * t);
+  static_cast<views::NativeWidgetViews*>(native_widget())->GetView()->SetTransform(transform_->Interpolate(t));
 }
 
 void KeyboardWidget::AnimationEnded(const ui::Animation* animation) {