diff --git a/partition.cpp b/partition.cpp
index 3957c654..76b558a0 100644
--- a/partition.cpp
+++ b/partition.cpp
@@ -1430,7 +1430,7 @@ bool TWPartition::Mount(bool Display_Error) {
 	}
 
 	string mount_fs = Current_File_System;
-	if (Current_File_System == "exfat" && TWFunc::Path_Exists("/sys/module/texfat"))
+	if (Current_File_System == "exfat")
 		mount_fs = "texfat";
 
 	if (!exfat_mounted &&
diff --git a/twrp-functions.cpp b/twrp-functions.cpp
index 5df44c69..49fcbaf9 100644
--- a/twrp-functions.cpp
+++ b/twrp-functions.cpp
@@ -901,7 +901,7 @@ void TWFunc::Fixup_Time_On_Boot()
 
 		gettimeofday(&tv, NULL);
 
-		if (tv.tv_sec > 1405209403) { // Anything older then 12 Jul 2014 23:56:43 GMT will do nicely thank you ;)
+		if (tv.tv_sec > 1513613049) { // Anything older then 12 Jul 2014 23:56:43 GMT will do nicely thank you ;)
 
 			LOGINFO("TWFunc::Fixup_Time: Date and time corrected: %s\n", TWFunc::Get_Current_Date().c_str());
 			return;