Index: www/lib/releases.php
===================================================================
--- www/lib/releases.php	(revision 1914)
+++ www/lib/releases.php	(working copy)
@@ -1660,9 +1660,23 @@
 		//
 		// Delete any parts and binaries which are older than the site's retention days
 		//
-		echo "Deleting parts which are older than ".$page->site->rawretentiondays." days\n";			
-		$db->query(sprintf("delete from parts where dateadded < %s - interval %d day", $db->escapeString($currTime_ori["now"]), $page->site->rawretentiondays));
+		echo "Deleting parts which are older than ".$page->site->rawretentiondays." days\n";
+		//$db->query(sprintf("delete from parts where dateadded < %s - interval %d day", $db->escapeString($currTime_ori["now"]), $page->site->rawretentiondays));
 
+		// Delete parts incrementally.  This is faster when there are tens or hundreds of million parts to be deleted.
+		$count = 0;
+		while (true)
+		{
+			$success = $db->query(sprintf("delete from parts where dateadded < %s - interval %d day limit 100000", $db->escapeString($currTime_ori["now"]), $page->site->rawretentiondays));
+			$num = mysql_affected_rows();
+        		if ($num == 0) {
+				echo "\n";
+                		break;
+        		}
+			$count += $num;
+			echo "\r  ".$count." rows deleted";
+		}
+
 		echo "Deleting binaries which are older than ".$page->site->rawretentiondays." days\n";			
 		$db->query(sprintf("delete from binaries where dateadded < %s - interval %d day", $db->escapeString($currTime_ori["now"]), $page->site->rawretentiondays));