diff --git a/app/sanity.c b/app/sanity.c
index 23b3d50d79..652da8873c 100644
--- a/app/sanity.c
+++ b/app/sanity.c
@@ -172,7 +172,7 @@ sanity_check_glib (void)
 {
 #define GLIB_REQUIRED_MAJOR 2
 #define GLIB_REQUIRED_MINOR 54
-#define GLIB_REQUIRED_MICRO 2
+#define GLIB_REQUIRED_MICRO 0
 
   const gchar *mismatch = glib_check_version (GLIB_REQUIRED_MAJOR,
                                               GLIB_REQUIRED_MINOR,
diff --git a/configure.ac b/configure.ac
index 04429b8adf..2d4dad5af6 100644
--- a/configure.ac
+++ b/configure.ac
@@ -45,7 +45,7 @@ m4_define([gimp_full_name], [GNU Image Manipulation Program])
 # required versions of other packages
 m4_define([babl_required_version], [0.1.38])
 m4_define([gegl_required_version], [0.3.27])
-m4_define([glib_required_version], [2.54.2])
+m4_define([glib_required_version], [2.54.0])
 m4_define([atk_required_version], [2.2.0])
 m4_define([gtk_required_version], [2.24.10])
 m4_define([gtk_win32_recommended_version], [2.24.32])