diff --git a/src/game/Spell.cpp b/src/game/Spell.cpp
index 8bea4c0..22bc036 100644
--- a/src/game/Spell.cpp
+++ b/src/game/Spell.cpp
@@ -804,10 +804,12 @@ void Spell::AddUnitTarget(Unit* pVictim, SpellEffectIndex effIndex)
         return;
 
     // Check for effect immune skip if immuned
-    bool immuned = pVictim->IsImmuneToSpellEffect(m_spellInfo, effIndex);
-
-    if (pVictim->GetTypeId() == TYPEID_UNIT && ((Creature*)pVictim)->IsTotem() && (m_spellFlags & SPELL_FLAG_REDIRECTED))
+    bool immuned;
+    if (pVictim->GetTypeId() == TYPEID_UNIT && ((Creature*)pVictim)->IsTotem() &&
+        (m_sender == pVictim || (m_spellFlags & SPELL_FLAG_REDIRECTED)))
         immuned = false;
+    else
+        immuned = pVictim->IsImmuneToSpellEffect(m_spellInfo, effIndex);
 
     ObjectGuid targetGUID = pVictim->GetObjectGuid();