diff --git a/src/engine/qcommon/md5.c b/src/engine/qcommon/md5.c
index 9c476ee..0fb79a6 100644
--- a/src/engine/qcommon/md5.c
+++ b/src/engine/qcommon/md5.c
@@ -370,26 +370,25 @@ static char *CalculateGUID(const char *key)
 
 char           *Com_MD5FileETCompat(const char *filename)
 {
-       char                       key[18];
+       char                       key[19] = {0};
        char              *buffer;
        int                len;
 
        len = FS_ReadFile(filename, (void **)&buffer);
        if (buffer)
        {
-               int i;
-               for (i = 0; i < 18; i++)
+               if (len >= 28)
                {
-                       key[i] = buffer[i + 10];
+                       int i;
+                       for (i = 0; i < 18; i++)
+                       {
+                               key[i] = buffer[i + 10];
+                       }
                }
-               key[18] = 0;
 
                return CalculateGUID(key);
        }
-       else
-       {
-               return 0;
-       }
+       return 0;
 }